{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "35e7cbed-1779-409f-bdc2-1599f14bb54a",
   "metadata": {},
   "source": [
    "Process raw Infinium 850k methylation array raw data using meffil R package: https://github.com/perishky/meffil/wiki/\n",
    "- this notebooks peforms QC and removal of bad quality samples\n",
    "- Date: 27.10.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ce3713-132f-4ec5-ae6a-c2adf2e8a7ac",
   "metadata": {},
   "source": [
    "### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fbc17c6e-d801-4ebd-b2ba-cd3abba5687e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "               _                           \n",
       "platform       x86_64-conda-linux-gnu      \n",
       "arch           x86_64                      \n",
       "os             linux-gnu                   \n",
       "system         x86_64, linux-gnu           \n",
       "status                                     \n",
       "major          4                           \n",
       "minor          5.1                         \n",
       "year           2025                        \n",
       "month          06                          \n",
       "day            13                          \n",
       "svn rev        88306                       \n",
       "language       R                           \n",
       "version.string R version 4.5.1 (2025-06-13)\n",
       "nickname       Great Square Root           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "R.version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1d38612-bba3-4b67-a305-38b3da353d13",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: illuminaio\n",
      "\n",
      "Loading required package: MASS\n",
      "\n",
      "Loading required package: lmtest\n",
      "\n",
      "Loading required package: zoo\n",
      "\n",
      "\n",
      "Attaching package: ‘zoo’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    yearmon, yearqtr\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.Date, as.Date.numeric\n",
      "\n",
      "\n",
      "Loading required package: sandwich\n",
      "\n",
      "Loading required package: sva\n",
      "\n",
      "Loading required package: mgcv\n",
      "\n",
      "Loading required package: nlme\n",
      "\n",
      "This is mgcv 1.9-3. For overview type 'help(\"mgcv-package\")'.\n",
      "\n",
      "Loading required package: genefilter\n",
      "\n",
      "\n",
      "Attaching package: ‘genefilter’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    area\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:vroom’:\n",
      "\n",
      "    spec\n",
      "\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Loading required package: plyr\n",
      "\n",
      "Loading required package: reshape2\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "Loading required package: knitr\n",
      "\n",
      "Loading required package: Cairo\n",
      "\n",
      "Loading required package: gridExtra\n",
      "\n",
      "Loading required package: markdown\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:genefilter’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "Loading required package: multcomp\n",
      "\n",
      "Loading required package: mvtnorm\n",
      "\n",
      "Loading required package: survival\n",
      "\n",
      "Loading required package: TH.data\n",
      "\n",
      "\n",
      "Attaching package: ‘TH.data’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    geyser\n",
      "\n",
      "\n",
      "Loading required package: lme4\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:tidyr’:\n",
      "\n",
      "    expand, pack, unpack\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘lme4’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    lmList\n",
      "\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "Loading required package: fastICA\n",
      "\n",
      "Loading required package: DNAcopy\n",
      "\n",
      "Loading required package: quadprog\n",
      "\n",
      "Loading required package: statmod\n",
      "\n",
      "Loading required package: gdsfmt\n",
      "\n",
      "Loading required package: SmartSVA\n",
      "\n",
      "Loading required package: isva\n",
      "\n",
      "Loading required package: qvalue\n",
      "\n",
      "Loading required package: JADE\n",
      "\n",
      "Loading required package: RSpectra\n",
      "\n",
      "Loading required package: preprocessCore\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise, summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## load libraries\n",
    "library(stringr)\n",
    "library(data.table) \n",
    "library(vroom)\n",
    "library(ggplot2)\n",
    "library(tidyr)\n",
    "library(limma)\n",
    "library(meffil)\n",
    "library(readxl)\n",
    "library(dplyr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "08604216-f69a-41a0-9a98-dd90d82f77df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set wd\n",
    "setwd('/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a4238a19-a058-4f36-b1e3-da2549d92fec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "40"
      ],
      "text/latex": [
       "40"
      ],
      "text/markdown": [
       "40"
      ],
      "text/plain": [
       "[1] 40"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# set # of cores\n",
    "library(parallel)\n",
    "cores = detectCores()\n",
    "cores\n",
    "options(mc.cores=cores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "63654e4d-513f-4e3f-a703-b868a3cfd78b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Sample_Name</th><th scope=col>Slide</th><th scope=col>sentrix_row</th><th scope=col>sentrix_col</th><th scope=col>Basename</th><th scope=col>Sex</th><th scope=col>batch</th><th scope=col>tissue.region</th><th scope=col>sample.wait.time</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>SD001/11B</td><td>200514040135</td><td>01</td><td>01</td><td>BrainSamples/data//idats_140716/200514040135_R01C01</td><td>M</td><td>2056G</td><td>Cortex</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>SD033/10 </td><td>200514040135</td><td>02</td><td>01</td><td>BrainSamples/data//idats_140716/200514040135_R02C01</td><td>M</td><td>2056G</td><td>Cortex</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>SD024/08 </td><td>200514040135</td><td>03</td><td>01</td><td>BrainSamples/data//idats_140716/200514040135_R03C01</td><td>M</td><td>2056G</td><td>Cortex</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>SD039/08 </td><td>200514040135</td><td>04</td><td>01</td><td>BrainSamples/data//idats_140716/200514040135_R04C01</td><td>M</td><td>2056G</td><td>Cortex</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>SD043/06 </td><td>200514040135</td><td>05</td><td>01</td><td>BrainSamples/data//idats_140716/200514040135_R05C01</td><td>F</td><td>2056G</td><td>Cortex</td><td>5</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>SD034/09B</td><td>200514040135</td><td>06</td><td>01</td><td>BrainSamples/data//idats_140716/200514040135_R06C01</td><td>M</td><td>2056G</td><td>Cortex</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & Sample\\_Name & Slide & sentrix\\_row & sentrix\\_col & Basename & Sex & batch & tissue.region & sample.wait.time\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & SD001/11B & 200514040135 & 01 & 01 & BrainSamples/data//idats\\_140716/200514040135\\_R01C01 & M & 2056G & Cortex & 5\\\\\n",
       "\t2 & SD033/10  & 200514040135 & 02 & 01 & BrainSamples/data//idats\\_140716/200514040135\\_R02C01 & M & 2056G & Cortex & 0\\\\\n",
       "\t3 & SD024/08  & 200514040135 & 03 & 01 & BrainSamples/data//idats\\_140716/200514040135\\_R03C01 & M & 2056G & Cortex & 5\\\\\n",
       "\t4 & SD039/08  & 200514040135 & 04 & 01 & BrainSamples/data//idats\\_140716/200514040135\\_R04C01 & M & 2056G & Cortex & 5\\\\\n",
       "\t5 & SD043/06  & 200514040135 & 05 & 01 & BrainSamples/data//idats\\_140716/200514040135\\_R05C01 & F & 2056G & Cortex & 5\\\\\n",
       "\t6 & SD034/09B & 200514040135 & 06 & 01 & BrainSamples/data//idats\\_140716/200514040135\\_R06C01 & M & 2056G & Cortex & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | Sample_Name &lt;chr&gt; | Slide &lt;chr&gt; | sentrix_row &lt;chr&gt; | sentrix_col &lt;chr&gt; | Basename &lt;chr&gt; | Sex &lt;chr&gt; | batch &lt;chr&gt; | tissue.region &lt;chr&gt; | sample.wait.time &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | SD001/11B | 200514040135 | 01 | 01 | BrainSamples/data//idats_140716/200514040135_R01C01 | M | 2056G | Cortex | 5 |\n",
       "| 2 | SD033/10  | 200514040135 | 02 | 01 | BrainSamples/data//idats_140716/200514040135_R02C01 | M | 2056G | Cortex | 0 |\n",
       "| 3 | SD024/08  | 200514040135 | 03 | 01 | BrainSamples/data//idats_140716/200514040135_R03C01 | M | 2056G | Cortex | 5 |\n",
       "| 4 | SD039/08  | 200514040135 | 04 | 01 | BrainSamples/data//idats_140716/200514040135_R04C01 | M | 2056G | Cortex | 5 |\n",
       "| 5 | SD043/06  | 200514040135 | 05 | 01 | BrainSamples/data//idats_140716/200514040135_R05C01 | F | 2056G | Cortex | 5 |\n",
       "| 6 | SD034/09B | 200514040135 | 06 | 01 | BrainSamples/data//idats_140716/200514040135_R06C01 | M | 2056G | Cortex | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Sample_Name Slide        sentrix_row sentrix_col Basename                                            Sex batch tissue.region sample.wait.time\n",
       "1 SD001/11B   200514040135 01          01          BrainSamples/data//idats_140716/200514040135_R01C01 M   2056G Cortex        5               \n",
       "2 SD033/10    200514040135 02          01          BrainSamples/data//idats_140716/200514040135_R02C01 M   2056G Cortex        0               \n",
       "3 SD024/08    200514040135 03          01          BrainSamples/data//idats_140716/200514040135_R03C01 M   2056G Cortex        5               \n",
       "4 SD039/08    200514040135 04          01          BrainSamples/data//idats_140716/200514040135_R04C01 M   2056G Cortex        5               \n",
       "5 SD043/06    200514040135 05          01          BrainSamples/data//idats_140716/200514040135_R05C01 F   2056G Cortex        5               \n",
       "6 SD034/09B   200514040135 06          01          BrainSamples/data//idats_140716/200514040135_R06C01 M   2056G Cortex        0               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>136</li><li>9</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 136\n",
       "\\item 9\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 136\n",
       "2. 9\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 136   9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# generate sample sheet\n",
    "samplesheet <- meffil.create.samplesheet('BrainSamples/data//idats_140716', recursive=TRUE)\n",
    "# update sex and Sample name\n",
    "batch <- fread('metadata/pheno.txt')\n",
    "\n",
    "upd <- \n",
    "fread('BrainSamples/data/idats_140716/Samples_Table_140716.csv') %>% \n",
    "    mutate(Sample_Name = paste0(`Sentrix Barcode`, \"_\", `Array`)) %>%\n",
    "    dplyr::select(c('Sample ID', 'Sample_Name')) %>%\n",
    "    dplyr::rename('Sample_Name2' = 'Sample ID')\n",
    "\n",
    "samplesheet <-\n",
    "inner_join(samplesheet, upd, by = 'Sample_Name') %>% \n",
    "    dplyr::mutate(Sample_Name = Sample_Name2) %>%\n",
    "    dplyr::select(-c(Sample_Name2)) %>%\n",
    "    dplyr::select(-c(Sex)) %>%\n",
    "    left_join(., batch, by = c('Sample_Name'='sample.ID')) %>%\n",
    "    dplyr::rename('Sex' = 'sex')\n",
    "samplesheet %>% head\n",
    "samplesheet %>% dim"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ff30d33-453f-4877-9f9b-bcaafcd2521c",
   "metadata": {},
   "source": [
    "### Quality Control"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bed83640-1237-4ab1-aa4f-f9a9fcb5b329",
   "metadata": {},
   "source": [
    "Next perform the background correction, dye bias correction, sex prediction and cell count estimates. The `meffil.qc` function processes your idat files and returns a `qc.object` for each sample. You need to specify which cell type reference you need. You can find cell type references with the `meffil.list.cell.type.references()` function. Currently there are whole blood and cord blood references implemented. \n",
    "- Set `cell.type.reference` to 'guintivano dlpfc' (brain tissue) \n",
    "- see: https://github.com/perishky/meffil/wiki/Sample-QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c316c5ba-f0fc-42ba-86ed-ea97b2adaa2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'andrews and bakulski cord blood'</li><li>'blood gse167998'</li><li>'blood gse35069'</li><li>'blood gse35069 chen'</li><li>'blood gse35069 complete'</li><li>'blood idoloptimized'</li><li>'blood idoloptimized epic'</li><li>'combined cord blood'</li><li>'cord blood gse68456'</li><li>'gervin and lyle cord blood'</li><li>'guintivano dlpfc'</li><li>'saliva gse147318'</li><li>'saliva gse48472'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'andrews and bakulski cord blood'\n",
       "\\item 'blood gse167998'\n",
       "\\item 'blood gse35069'\n",
       "\\item 'blood gse35069 chen'\n",
       "\\item 'blood gse35069 complete'\n",
       "\\item 'blood idoloptimized'\n",
       "\\item 'blood idoloptimized epic'\n",
       "\\item 'combined cord blood'\n",
       "\\item 'cord blood gse68456'\n",
       "\\item 'gervin and lyle cord blood'\n",
       "\\item 'guintivano dlpfc'\n",
       "\\item 'saliva gse147318'\n",
       "\\item 'saliva gse48472'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'andrews and bakulski cord blood'\n",
       "2. 'blood gse167998'\n",
       "3. 'blood gse35069'\n",
       "4. 'blood gse35069 chen'\n",
       "5. 'blood gse35069 complete'\n",
       "6. 'blood idoloptimized'\n",
       "7. 'blood idoloptimized epic'\n",
       "8. 'combined cord blood'\n",
       "9. 'cord blood gse68456'\n",
       "10. 'gervin and lyle cord blood'\n",
       "11. 'guintivano dlpfc'\n",
       "12. 'saliva gse147318'\n",
       "13. 'saliva gse48472'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"andrews and bakulski cord blood\" \"blood gse167998\"                 \"blood gse35069\"                  \"blood gse35069 chen\"             \"blood gse35069 complete\"        \n",
       " [6] \"blood idoloptimized\"             \"blood idoloptimized epic\"        \"combined cord blood\"             \"cord blood gse68456\"             \"gervin and lyle cord blood\"     \n",
       "[11] \"guintivano dlpfc\"                \"saliva gse147318\"                \"saliva gse48472\"                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meffil.list.cell.type.references()\n",
    "qc.objects <- meffil.qc(samplesheet, cell.type.reference='guintivano dlpfc', verbose=FALSE)\n",
    "save(qc.objects,file=\"meffil_data/qc.objects.Robj\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ada35f9-f3f0-4ddd-86e6-f6bcd1b3a93a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'SD001/11B'</li><li>'SD033/10'</li><li>'SD024/08'</li><li>'SD039/08'</li><li>'SD043/06'</li><li>'SD034/09B'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'SD001/11B'\n",
       "\\item 'SD033/10'\n",
       "\\item 'SD024/08'\n",
       "\\item 'SD039/08'\n",
       "\\item 'SD043/06'\n",
       "\\item 'SD034/09B'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'SD001/11B'\n",
       "2. 'SD033/10'\n",
       "3. 'SD024/08'\n",
       "4. 'SD039/08'\n",
       "5. 'SD043/06'\n",
       "6. 'SD034/09B'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"SD001/11B\" \"SD033/10\"  \"SD024/08\"  \"SD039/08\"  \"SD043/06\"  \"SD034/09B\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "136"
      ],
      "text/latex": [
       "136"
      ],
      "text/markdown": [
       "136"
      ],
      "text/plain": [
       "[1] 136"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(qc.objects) %>% head\n",
    "length(qc.objects)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "460230f7-7879-4ab2-a1d8-735a9cea46c0",
   "metadata": {},
   "source": [
    "### Remove samples with low genotype concordance "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6665c85-6665-4304-a674-e9268093e7eb",
   "metadata": {},
   "source": [
    "- If you have genotype data available on the same individuals with methylation profiles you can check for ID mismatches. The methylation arrays have 65 SNPs which can be extracted from the methylation data. These 65 SNPs can be compared to genotypes measured with genotype arrays.\n",
    "- see `meffil_00_match_samples.ipynb` for details, where the genotype concordance was calculated "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "782c5b48-e353-4646-aa52-24ac1c5461d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load meffil objects\n",
    "# load('meffil_data/qc.objects.Robj')\n",
    "# load('meffil_data/qcsummary.Robj')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e555d053-da28-4d57-a90f-cac38109132a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10 samples removed due to low genotype concordance: SD010/09 SD023/11B SD023/13 SD030/09 SD032/08 SD036/10 SD036/13 SD036/14B SD038/08 SD042/13 \n",
      "126 samples remaining"
     ]
    }
   ],
   "source": [
    "### remove low concordance samples < 0.75\n",
    "genconc <- fread('meffil_data/sampleID.mapping.genconc.txt')\n",
    "bad.conc.samples <- genconc %>% filter( gen.concordance < 0.75) %>% pull(sample.ID)\n",
    "cat(length(bad.conc.samples), 'samples removed due to low genotype concordance:', bad.conc.samples, '\\n')\n",
    "\n",
    "qc.objects.conc <- meffil.remove.samples(qc.objects, bad.conc.samples)\n",
    "cat(length(qc.objects.conc), 'samples remaining')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0047a6fe-8c0d-4fee-94e2-958406ebf85c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'epic'"
      ],
      "text/latex": [
       "'epic'"
      ],
      "text/markdown": [
       "'epic'"
      ],
      "text/plain": [
       "[1] \"epic\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "featureset <- qc.objects[[2]]$featureset\n",
    "featureset\n",
    "#writeLines(meffil.snp.names(featureset), con=\"snp-names.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d24f6df1-5e7c-4bd5-974e-6b332feaba3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genotypes extracted using:\n",
    "# >plink2 --pfile genotypingdata/plink_files/pgen/imputed_allchr_newIDs --extract snp-names-newID.txt --recode A --out meffil_data/genotypes-imp-newIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bd629c0e-855b-44c1-a8f3-47d04ca8e91a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 126 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>SD001/11B</th><th scope=col>SD033/10</th><th scope=col>SD024/08</th><th scope=col>SD039/08</th><th scope=col>SD043/06</th><th scope=col>SD034/09B</th><th scope=col>SD025/13</th><th scope=col>SD027/11</th><th scope=col>SD004/06</th><th scope=col>SD025/09</th><th scope=col>⋯</th><th scope=col>SD024/14B</th><th scope=col>SD008/09</th><th scope=col>SD032/09</th><th scope=col>SD022/08B</th><th scope=col>SD048/12</th><th scope=col>SD055/12</th><th scope=col>SD036/12</th><th scope=col>SD033/08</th><th scope=col>SD025/08</th><th scope=col>SD031/09</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>rs3936238</th><td>0</td><td>0</td><td>1</td><td>1</td><td>1</td><td>2</td><td>1</td><td>0</td><td>0</td><td>1</td><td>⋯</td><td>1</td><td>0</td><td>1</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>rs877309</th><td>1</td><td>1</td><td>2</td><td>0</td><td>0</td><td>2</td><td>1</td><td>2</td><td>1</td><td>0</td><td>⋯</td><td>1</td><td>1</td><td>2</td><td>0</td><td>1</td><td>2</td><td>1</td><td>0</td><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>rs213028</th><td>0</td><td>1</td><td>2</td><td>2</td><td>2</td><td>1</td><td>1</td><td>2</td><td>2</td><td>2</td><td>⋯</td><td>2</td><td>1</td><td>1</td><td>2</td><td>2</td><td>2</td><td>1</td><td>1</td><td>2</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>rs11249206</th><td>1</td><td>1</td><td>0</td><td>1</td><td>0</td><td>0</td><td>2</td><td>1</td><td>1</td><td>1</td><td>⋯</td><td>2</td><td>1</td><td>1</td><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>rs654498</th><td>0</td><td>1</td><td>0</td><td>1</td><td>0</td><td>0</td><td>0</td><td>1</td><td>0</td><td>0</td><td>⋯</td><td>1</td><td>0</td><td>1</td><td>2</td><td>2</td><td>2</td><td>2</td><td>0</td><td>1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>rs715359</th><td>1</td><td>1</td><td>1</td><td>0</td><td>2</td><td>0</td><td>1</td><td>2</td><td>2</td><td>1</td><td>⋯</td><td>2</td><td>2</td><td>1</td><td>1</td><td>2</td><td>2</td><td>1</td><td>1</td><td>1</td><td>2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 126 of type int\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & SD001/11B & SD033/10 & SD024/08 & SD039/08 & SD043/06 & SD034/09B & SD025/13 & SD027/11 & SD004/06 & SD025/09 & ⋯ & SD024/14B & SD008/09 & SD032/09 & SD022/08B & SD048/12 & SD055/12 & SD036/12 & SD033/08 & SD025/08 & SD031/09\\\\\n",
       "\\hline\n",
       "\trs3936238 & 0 & 0 & 1 & 1 & 1 & 2 & 1 & 0 & 0 & 1 & ⋯ & 1 & 0 & 1 & 1 & 0 & 0 & 0 & 0 & 1 & 1\\\\\n",
       "\trs877309 & 1 & 1 & 2 & 0 & 0 & 2 & 1 & 2 & 1 & 0 & ⋯ & 1 & 1 & 2 & 0 & 1 & 2 & 1 & 0 & 1 & 1\\\\\n",
       "\trs213028 & 0 & 1 & 2 & 2 & 2 & 1 & 1 & 2 & 2 & 2 & ⋯ & 2 & 1 & 1 & 2 & 2 & 2 & 1 & 1 & 2 & 1\\\\\n",
       "\trs11249206 & 1 & 1 & 0 & 1 & 0 & 0 & 2 & 1 & 1 & 1 & ⋯ & 2 & 1 & 1 & 0 & 1 & 1 & 1 & 1 & 0 & 0\\\\\n",
       "\trs654498 & 0 & 1 & 0 & 1 & 0 & 0 & 0 & 1 & 0 & 0 & ⋯ & 1 & 0 & 1 & 2 & 2 & 2 & 2 & 0 & 1 & 0\\\\\n",
       "\trs715359 & 1 & 1 & 1 & 0 & 2 & 0 & 1 & 2 & 2 & 1 & ⋯ & 2 & 2 & 1 & 1 & 2 & 2 & 1 & 1 & 1 & 2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 126 of type int\n",
       "\n",
       "| <!--/--> | SD001/11B | SD033/10 | SD024/08 | SD039/08 | SD043/06 | SD034/09B | SD025/13 | SD027/11 | SD004/06 | SD025/09 | ⋯ | SD024/14B | SD008/09 | SD032/09 | SD022/08B | SD048/12 | SD055/12 | SD036/12 | SD033/08 | SD025/08 | SD031/09 |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| rs3936238 | 0 | 0 | 1 | 1 | 1 | 2 | 1 | 0 | 0 | 1 | ⋯ | 1 | 0 | 1 | 1 | 0 | 0 | 0 | 0 | 1 | 1 |\n",
       "| rs877309 | 1 | 1 | 2 | 0 | 0 | 2 | 1 | 2 | 1 | 0 | ⋯ | 1 | 1 | 2 | 0 | 1 | 2 | 1 | 0 | 1 | 1 |\n",
       "| rs213028 | 0 | 1 | 2 | 2 | 2 | 1 | 1 | 2 | 2 | 2 | ⋯ | 2 | 1 | 1 | 2 | 2 | 2 | 1 | 1 | 2 | 1 |\n",
       "| rs11249206 | 1 | 1 | 0 | 1 | 0 | 0 | 2 | 1 | 1 | 1 | ⋯ | 2 | 1 | 1 | 0 | 1 | 1 | 1 | 1 | 0 | 0 |\n",
       "| rs654498 | 0 | 1 | 0 | 1 | 0 | 0 | 0 | 1 | 0 | 0 | ⋯ | 1 | 0 | 1 | 2 | 2 | 2 | 2 | 0 | 1 | 0 |\n",
       "| rs715359 | 1 | 1 | 1 | 0 | 2 | 0 | 1 | 2 | 2 | 1 | ⋯ | 2 | 2 | 1 | 1 | 2 | 2 | 1 | 1 | 1 | 2 |\n",
       "\n"
      ],
      "text/plain": [
       "           SD001/11B SD033/10 SD024/08 SD039/08 SD043/06 SD034/09B SD025/13 SD027/11 SD004/06 SD025/09 ⋯ SD024/14B SD008/09 SD032/09 SD022/08B SD048/12 SD055/12 SD036/12 SD033/08 SD025/08 SD031/09\n",
       "rs3936238  0         0        1        1        1        2         1        0        0        1        ⋯ 1         0        1        1         0        0        0        0        1        1       \n",
       "rs877309   1         1        2        0        0        2         1        2        1        0        ⋯ 1         1        2        0         1        2        1        0        1        1       \n",
       "rs213028   0         1        2        2        2        1         1        2        2        2        ⋯ 2         1        1        2         2        2        1        1        2        1       \n",
       "rs11249206 1         1        0        1        0        0         2        1        1        1        ⋯ 2         1        1        0         1        1        1        1        0        0       \n",
       "rs654498   0         1        0        1        0        0         0        1        0        0        ⋯ 1         0        1        2         2        2        2        0        1        0       \n",
       "rs715359   1         1        1        0        2        0         1        2        2        1        ⋯ 2         2        1        1         2        2        1        1        1        2       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## load genotypes (imputed)\n",
    "genotypes0 <- meffil.extract.genotypes(\"meffil_data/genotypes-imp-newIDs.raw\")\n",
    "genotypes0 <- genotypes0[,match(names(qc.objects.conc), colnames(genotypes0))]\n",
    "genotypes_df <- \n",
    "    as.data.frame(genotypes0) %>%\n",
    "    tibble::rownames_to_column(\"gen.id\")\n",
    "## fix SNP names \n",
    "rsids <-\n",
    "fread('snp-names-pvar-table.txt') %>% \n",
    "    select(snp, gen.id) %>% \n",
    "    mutate(\n",
    "        gen.id = ifelse(str_detect(gen.id, \"^X:\"), gen.id, paste0(\"X\", gen.id)),\n",
    "        gen.id = gsub(\":\", \".\", gen.id)\n",
    "          )\n",
    "genotypes_df2 <- left_join(genotypes_df, rsids, by = 'gen.id') %>% select(-c(gen.id))\n",
    "genotypes     <- as.matrix(genotypes_df2[,!(names(genotypes_df2) %in% 'snp')])\n",
    "rownames(genotypes) <- genotypes_df2$snp  \n",
    "genotypes %>% head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9944596c-4bed-48c3-a339-bcba1806f8ba",
   "metadata": {},
   "source": [
    "### Generate QC report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "764243c4-3440-4c05-978e-0c279eeaa806",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc.parameters <- \n",
    "meffil.qc.parameters(\n",
    "    meth.unmeth.outlier.sd                = 5,\n",
    "\tbeadnum.samples.threshold             = 0.1,\n",
    "\tdetectionp.samples.threshold          = 0.1,\n",
    "\tdetectionp.cpgs.threshold             = 0.1, \n",
    "\tbeadnum.cpgs.threshold                = 0.1,\n",
    "\tsex.outlier.sd                        = 5,\n",
    "\tsnp.concordance.threshold             = 0.95,\n",
    "\tsample.genotype.concordance.threshold = 0.75\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4cf210b7-639b-4179-b106-0b17a4cca920",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[meffil.qc.summary] Tue Nov  4 09:54:03 2025 Sex summary TRUE \n",
      "[meffil.qc.summary] Tue Nov  4 09:54:03 2025 Meth vs unmeth summary \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe `<scale>` argument of `guides()` cannot be `FALSE`. Use \"none\" instead as of ggplot2 3.3.4.\n",
      "\u001b[36mℹ\u001b[39m The deprecated feature was likely used in the \u001b[34mmeffil\u001b[39m package.\n",
      "  Please report the issue to the authors.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[meffil.qc.summary] Tue Nov  4 09:54:03 2025 Control means summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:54:03 2025 Sample detection summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:54:20 2025 CpG detection summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:54:30 2025 Sample bead numbers summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:54:47 2025 CpG bead numbers summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:54:49 2025 Cell count summary \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe `fun.y` argument of `stat_summary()` is deprecated as of ggplot2 3.3.0.\n",
      "\u001b[36mℹ\u001b[39m Please use the `fun` argument instead.\n",
      "\u001b[36mℹ\u001b[39m The deprecated feature was likely used in the \u001b[34mmeffil\u001b[39m package.\n",
      "  Please report the issue to the authors.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[meffil.qc.summary] Tue Nov  4 09:54:49 2025 Genotype concordance \n"
     ]
    }
   ],
   "source": [
    "## full dataset\n",
    "qc.summary <- meffil.qc.summary(\n",
    "\tqc.objects,\n",
    "\tparameters = qc.parameters,\n",
    "\tgenotypes=genotypes\n",
    ")\n",
    "\n",
    "save(qc.summary, file=\"meffil_data/qcsummary.Robj\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c0ec958a-3944-4686-8349-2e5df85d611b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[meffil.qc.report] Tue Nov  4 09:55:22 2025 Writing report as html file to meffil_data/qc-report.html \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/qc-report.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/38                   \n",
      "2/38 [unnamed-chunk-1] \n",
      "3/38                   \n",
      "4/38 [unnamed-chunk-2] \n",
      "5/38                   \n",
      "6/38 [unnamed-chunk-3] \n",
      "7/38                   \n",
      "8/38 [unnamed-chunk-4] \n",
      "9/38                   \n",
      "10/38 [unnamed-chunk-5] \n",
      "11/38                   \n",
      "12/38 [unnamed-chunk-6] \n",
      "13/38                   \n",
      "14/38 [unnamed-chunk-7] \n",
      "15/38                   \n",
      "16/38 [unnamed-chunk-8] \n",
      "17/38                   \n",
      "18/38 [unnamed-chunk-9] \n",
      "19/38                   \n",
      "20/38 [unnamed-chunk-10]\n",
      "21/38                   \n",
      "22/38 [unnamed-chunk-11]\n",
      "23/38                   \n",
      "24/38 [unnamed-chunk-12]\n",
      "25/38                   \n",
      "26/38 [unnamed-chunk-13]\n",
      "27/38                   \n",
      "28/38 [unnamed-chunk-14]\n",
      "29/38                   \n",
      "30/38 [unnamed-chunk-15]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/cell-counts.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/4                   \n",
      "2/4 [unnamed-chunk-35]\n",
      "3/4                   \n",
      "4/4 [unnamed-chunk-36]\n",
      "31/38                   \n",
      "32/38 [unnamed-chunk-16]\n",
      "33/38                   \n",
      "34/38 [unnamed-chunk-17]\n",
      "35/38                   \n",
      "36/38 [unnamed-chunk-18]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/genotype-concordance.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/9                   \n",
      "2/9 [unnamed-chunk-42]\n",
      "3/9                   \n",
      "4/9 [unnamed-chunk-43]\n",
      "5/9                   \n",
      "6/9 [unnamed-chunk-44]\n",
      "7/9                   \n",
      "8/9 [unnamed-chunk-45]\n",
      "9/9                   \n",
      "37/38                   \n",
      "38/38 [unnamed-chunk-19]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "output file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.md\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.html'"
      ],
      "text/latex": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB\\_methylation/meffil\\_data/qc-report.html'"
      ],
      "text/markdown": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.html'"
      ],
      "text/plain": [
       "[1] \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.html\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meffil.qc.report(qc.summary, output.file=\"meffil_data/qc-report.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f2f51323-2475-4765-b860-36a45800a163",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[meffil.qc.summary] Tue Nov  4 09:55:54 2025 Sex summary TRUE \n",
      "[meffil.qc.summary] Tue Nov  4 09:55:54 2025 Meth vs unmeth summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:55:54 2025 Control means summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:55:54 2025 Sample detection summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:56:10 2025 CpG detection summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:56:10 2025 Sample bead numbers summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:56:25 2025 CpG bead numbers summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:56:27 2025 Cell count summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:56:27 2025 Genotype concordance \n"
     ]
    }
   ],
   "source": [
    "## high gen.conc dataset\n",
    "qc.summary <- meffil.qc.summary(\n",
    "\tqc.objects.conc,\n",
    "\tparameters = qc.parameters,\n",
    "\tgenotypes=genotypes\n",
    ")\n",
    "\n",
    "save(qc.summary, file=\"meffil_data/qcsummary.highconc.Robj\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "43f25fbc-ec72-4bda-a333-4fed79d9ef08",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[meffil.qc.report] Tue Nov  4 09:56:40 2025 Writing report as html file to meffil_data/qc-report.highconc.html \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/qc-report.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/38                   \n",
      "2/38 [unnamed-chunk-1] \n",
      "3/38                   \n",
      "4/38 [unnamed-chunk-2] \n",
      "5/38                   \n",
      "6/38 [unnamed-chunk-3] \n",
      "7/38                   \n",
      "8/38 [unnamed-chunk-4] \n",
      "9/38                   \n",
      "10/38 [unnamed-chunk-5] \n",
      "11/38                   \n",
      "12/38 [unnamed-chunk-6] \n",
      "13/38                   \n",
      "14/38 [unnamed-chunk-7] \n",
      "15/38                   \n",
      "16/38 [unnamed-chunk-8] \n",
      "17/38                   \n",
      "18/38 [unnamed-chunk-9] \n",
      "19/38                   \n",
      "20/38 [unnamed-chunk-10]\n",
      "21/38                   \n",
      "22/38 [unnamed-chunk-11]\n",
      "23/38                   \n",
      "24/38 [unnamed-chunk-12]\n",
      "25/38                   \n",
      "26/38 [unnamed-chunk-13]\n",
      "27/38                   \n",
      "28/38 [unnamed-chunk-14]\n",
      "29/38                   \n",
      "30/38 [unnamed-chunk-15]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/cell-counts.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/4                   \n",
      "2/4 [unnamed-chunk-35]\n",
      "3/4                   \n",
      "4/4 [unnamed-chunk-36]\n",
      "31/38                   \n",
      "32/38 [unnamed-chunk-16]\n",
      "33/38                   \n",
      "34/38 [unnamed-chunk-17]\n",
      "35/38                   \n",
      "36/38 [unnamed-chunk-18]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/genotype-concordance.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/9                   \n",
      "2/9 [unnamed-chunk-42]\n",
      "3/9                   \n",
      "4/9 [unnamed-chunk-43]\n",
      "5/9                   \n",
      "6/9 [unnamed-chunk-44]\n",
      "7/9                   \n",
      "8/9 [unnamed-chunk-45]\n",
      "9/9                   \n",
      "37/38                   \n",
      "38/38 [unnamed-chunk-19]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "output file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.highconc.md\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.highconc.html'"
      ],
      "text/latex": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB\\_methylation/meffil\\_data/qc-report.highconc.html'"
      ],
      "text/markdown": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.highconc.html'"
      ],
      "text/plain": [
       "[1] \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.highconc.html\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meffil.qc.report(qc.summary, output.file=\"meffil_data/qc-report.highconc.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a46ddaac-d08b-4d69-af2b-83d9e8f1227d",
   "metadata": {},
   "source": [
    "### Remove bad samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e860abdd-64b4-4ac4-8a13-9c45e014afdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### qc parameters\n",
    "qc.parameters <- \n",
    "meffil.qc.parameters(\n",
    "    meth.unmeth.outlier.sd                = 5,\n",
    "\tbeadnum.samples.threshold             = 0.1,\n",
    "\tdetectionp.samples.threshold          = 0.1,\n",
    "\tdetectionp.cpgs.threshold             = 0.1, \n",
    "\tbeadnum.cpgs.threshold                = 0.1,\n",
    "\tsex.outlier.sd                        = 5,\n",
    "\tsnp.concordance.threshold             = 0.9,\n",
    "\tsample.genotype.concordance.threshold = 0.75\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d8386a97-b868-4b52-92e5-bbf489c72e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load meffil objects\n",
    "# load('meffil_data/qc.objects.Robj')\n",
    "# load('meffil_data/qcsummary.Robj')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e3bef177-edae-442d-b657-44c75d9758e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Control probe (spec1.ratio) \n",
       "                          1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## check\n",
    "outlier <- qc.summary$bad.samples\n",
    "table(outlier$issue)\n",
    "index <- outlier$issue %in% c(\"Control probe (dye.bias)\", \n",
    "                              \"Methylated vs Unmethylated\",\n",
    "                              \"X-Y ratio outlier\",\n",
    "                              \"Low bead numbers\",\n",
    "                              \"Detection p-value\",\n",
    "                              \"Sex mismatch\",\n",
    "                              \"Genotype mismatch\",\n",
    "                              \"Control probe (bisulfite1)\",\n",
    "                              \"Control probe (bisulfite2)\")\n",
    "\n",
    "outlier <- outlier[index,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bc35016b-8e58-43a1-9c7a-a59ec8b13c3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of samples before QC: 136 \n",
      "# of samples after genotype concordance QC: 126 \n",
      "# of samples after full QC: 126"
     ]
    }
   ],
   "source": [
    "cat('# of samples before QC:', length(qc.objects), '\\n')\n",
    "cat('# of samples after genotype concordance QC:', length(qc.objects.conc), '\\n')\n",
    "qc.objects <- meffil.remove.samples(qc.objects.conc, outlier$sample.name)\n",
    "cat('# of samples after full QC:', length(qc.objects))\n",
    "save(qc.objects,file=\"meffil_data/qc.objects.clean.Robj\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "853f9ea2-fc2f-443e-b3df-d015fe7ebfbf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[meffil.qc.summary] Tue Nov  4 09:57:45 2025 Sex summary TRUE \n",
      "[meffil.qc.summary] Tue Nov  4 09:57:45 2025 Meth vs unmeth summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:57:45 2025 Control means summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:57:45 2025 Sample detection summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:58:00 2025 CpG detection summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:58:01 2025 Sample bead numbers summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:58:17 2025 CpG bead numbers summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:58:18 2025 Cell count summary \n",
      "[meffil.qc.summary] Tue Nov  4 09:58:18 2025 Genotype concordance \n"
     ]
    }
   ],
   "source": [
    "### Rerun QC summary on clean dataset\n",
    "qc.summary <- meffil.qc.summary(qc.objects, parameters=qc.parameters, genotypes = genotypes)\n",
    "save(qc.summary, file = \"meffil_data/qcsummary.clean.Robj\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "677046c6-352e-4b8c-b83e-bbbc8fb0a8b6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[meffil.qc.report] Tue Nov  4 09:58:31 2025 Writing report as html file to meffil_data/qc-report.clean.html \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/qc-report.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/38                   \n",
      "2/38 [unnamed-chunk-1] \n",
      "3/38                   \n",
      "4/38 [unnamed-chunk-2] \n",
      "5/38                   \n",
      "6/38 [unnamed-chunk-3] \n",
      "7/38                   \n",
      "8/38 [unnamed-chunk-4] \n",
      "9/38                   \n",
      "10/38 [unnamed-chunk-5] \n",
      "11/38                   \n",
      "12/38 [unnamed-chunk-6] \n",
      "13/38                   \n",
      "14/38 [unnamed-chunk-7] \n",
      "15/38                   \n",
      "16/38 [unnamed-chunk-8] \n",
      "17/38                   \n",
      "18/38 [unnamed-chunk-9] \n",
      "19/38                   \n",
      "20/38 [unnamed-chunk-10]\n",
      "21/38                   \n",
      "22/38 [unnamed-chunk-11]\n",
      "23/38                   \n",
      "24/38 [unnamed-chunk-12]\n",
      "25/38                   \n",
      "26/38 [unnamed-chunk-13]\n",
      "27/38                   \n",
      "28/38 [unnamed-chunk-14]\n",
      "29/38                   \n",
      "30/38 [unnamed-chunk-15]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/cell-counts.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/4                   \n",
      "2/4 [unnamed-chunk-35]\n",
      "3/4                   \n",
      "4/4 [unnamed-chunk-36]\n",
      "31/38                   \n",
      "32/38 [unnamed-chunk-16]\n",
      "33/38                   \n",
      "34/38 [unnamed-chunk-17]\n",
      "35/38                   \n",
      "36/38 [unnamed-chunk-18]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "processing file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/R-packages/meffil/reports/genotype-concordance.rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/9                   \n",
      "2/9 [unnamed-chunk-42]\n",
      "3/9                   \n",
      "4/9 [unnamed-chunk-43]\n",
      "5/9                   \n",
      "6/9 [unnamed-chunk-44]\n",
      "7/9                   \n",
      "8/9 [unnamed-chunk-45]\n",
      "9/9                   \n",
      "37/38                   \n",
      "38/38 [unnamed-chunk-19]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "output file: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.clean.md\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.clean.html'"
      ],
      "text/latex": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB\\_methylation/meffil\\_data/qc-report.clean.html'"
      ],
      "text/markdown": [
       "'/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.clean.html'"
      ],
      "text/plain": [
       "[1] \"/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/meffil_data/qc-report.clean.html\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## new qc report\n",
    "meffil.qc.report(qc.summary, output.file=\"meffil_data/qc-report.clean.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7546b387-ca45-4324-8e6c-48b82ccce669",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.5.1 (2025-06-13)\n",
       "Platform: x86_64-conda-linux-gnu\n",
       "Running under: Rocky Linux 9.5 (Blue Onyx)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/jupiter-setup/envs/jpt/lib/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8      \n",
       " [8] LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/London\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] dplyr_1.1.4           readxl_1.4.5          meffil_1.5.1          preprocessCore_1.70.0 SmartSVA_0.1.3        RSpectra_0.16-2       isva_1.9              JADE_2.0-4           \n",
       " [9] qvalue_2.40.0         gdsfmt_1.44.1         statmod_1.5.1         quadprog_1.5-8        DNAcopy_1.82.0        fastICA_1.2-7         lme4_1.1-37           Matrix_1.7-4         \n",
       "[17] multcomp_1.4-29       TH.data_1.1-4         survival_3.8-3        mvtnorm_1.3-3         matrixStats_1.5.0     markdown_2.0          gridExtra_2.3         Cairo_1.6-5          \n",
       "[25] knitr_1.50            reshape2_1.4.4        plyr_1.8.9            sva_3.56.0            BiocParallel_1.42.2   genefilter_1.90.0     mgcv_1.9-3            nlme_3.1-168         \n",
       "[33] sandwich_3.1-1        lmtest_0.9-40         zoo_1.8-14            MASS_7.3-65           illuminaio_0.50.0     limma_3.64.3          tidyr_1.3.1           ggplot2_4.0.0        \n",
       "[41] vroom_1.6.6           data.table_1.17.8     stringr_1.5.2        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] RColorBrewer_1.1-3      jsonlite_2.0.0          magrittr_2.0.4          farver_2.1.2            nloptr_2.2.1            vctrs_0.6.5             memoise_2.0.1           minqa_1.2.8            \n",
       " [9] askpass_1.2.1           base64enc_0.1-3         htmltools_0.5.8.1       cellranger_1.1.0        cachem_1.1.0            uuid_1.2-1              commonmark_2.0.0        mime_0.13              \n",
       "[17] lifecycle_1.0.4         pkgconfig_2.0.3         R6_2.6.1                fastmap_1.2.0           GenomeInfoDbData_1.2.14 rbibutils_2.3           MatrixGenerics_1.20.0   clue_0.3-66            \n",
       "[25] digest_0.6.37           AnnotationDbi_1.70.0    S4Vectors_0.46.0        RSQLite_2.4.3           base64_2.0.2            labeling_0.4.3          httr_1.4.7              compiler_4.5.1         \n",
       "[33] bit64_4.6.0-1           withr_3.0.2             S7_0.2.0                DBI_1.2.3               openssl_2.3.4           tools_4.5.1             glue_1.8.0              grid_4.5.1             \n",
       "[41] pbdZMQ_0.3-14           cluster_2.1.8.1         generics_0.1.4          gtable_0.3.6            tzdb_0.5.0              XVector_0.48.0          BiocGenerics_0.54.1     pillar_1.11.1          \n",
       "[49] IRdisplay_1.1           splines_4.5.1           lattice_0.22-7          bit_4.6.0               annotate_1.86.1         tidyselect_1.2.1        locfit_1.5-9.12         Biostrings_2.76.0      \n",
       "[57] reformulas_0.4.1        litedown_0.7            IRanges_2.42.0          edgeR_4.6.3             stats4_4.5.1            xfun_0.53               Biobase_2.68.0          stringi_1.8.7          \n",
       "[65] UCSC.utils_1.4.0        boot_1.3-32             evaluate_1.0.5          codetools_0.2-20        tibble_3.3.0            cli_3.6.5               IRkernel_1.3.2          xtable_1.8-4           \n",
       "[73] Rdpack_2.6.4            repr_1.1.7              dichromat_2.0-0.1       Rcpp_1.1.0              GenomeInfoDb_1.44.3     png_0.1-8               XML_3.99-0.19           blob_1.2.4             \n",
       "[81] scales_1.4.0            purrr_1.1.0             crayon_1.5.3            rlang_1.1.6             KEGGREST_1.48.1        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
