{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "35e7cbed-1779-409f-bdc2-1599f14bb54a",
   "metadata": {},
   "source": [
    "Make phenotype and covariate files for meQTL using tensorQTL, see: https://github.com/broadinstitute/tensorqtl?tab=readme-ov-file\n",
    "- Date: 06.11.25\n",
    "- UPDATE 05.01.26: run on INT phenotypes "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ce3713-132f-4ec5-ae6a-c2adf2e8a7ac",
   "metadata": {},
   "source": [
    "### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fbc17c6e-d801-4ebd-b2ba-cd3abba5687e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "               _                           \n",
       "platform       x86_64-conda-linux-gnu      \n",
       "arch           x86_64                      \n",
       "os             linux-gnu                   \n",
       "system         x86_64, linux-gnu           \n",
       "status                                     \n",
       "major          4                           \n",
       "minor          5.1                         \n",
       "year           2025                        \n",
       "month          06                          \n",
       "day            13                          \n",
       "svn rev        88306                       \n",
       "language       R                           \n",
       "version.string R version 4.5.1 (2025-06-13)\n",
       "nickname       Great Square Root           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "R.version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1d38612-bba3-4b67-a305-38b3da353d13",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: illuminaio\n",
      "\n",
      "Loading required package: MASS\n",
      "\n",
      "Loading required package: lmtest\n",
      "\n",
      "Loading required package: zoo\n",
      "\n",
      "\n",
      "Attaching package: ‘zoo’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    yearmon, yearqtr\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.Date, as.Date.numeric\n",
      "\n",
      "\n",
      "Loading required package: sandwich\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "Loading required package: sva\n",
      "\n",
      "Loading required package: mgcv\n",
      "\n",
      "Loading required package: nlme\n",
      "\n",
      "This is mgcv 1.9-4. For overview type '?mgcv'.\n",
      "\n",
      "Loading required package: genefilter\n",
      "\n",
      "\n",
      "Attaching package: ‘genefilter’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    area\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:vroom’:\n",
      "\n",
      "    spec\n",
      "\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Loading required package: plyr\n",
      "\n",
      "Loading required package: reshape2\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "Loading required package: knitr\n",
      "\n",
      "Loading required package: Cairo\n",
      "\n",
      "Loading required package: gridExtra\n",
      "\n",
      "Loading required package: markdown\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:plyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:genefilter’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "Loading required package: multcomp\n",
      "\n",
      "Loading required package: mvtnorm\n",
      "\n",
      "Loading required package: survival\n",
      "\n",
      "Loading required package: TH.data\n",
      "\n",
      "\n",
      "Attaching package: ‘TH.data’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    geyser\n",
      "\n",
      "\n",
      "Loading required package: lme4\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "\n",
      "Attaching package: ‘lme4’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    lmList\n",
      "\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "Loading required package: fastICA\n",
      "\n",
      "Loading required package: DNAcopy\n",
      "\n",
      "Loading required package: quadprog\n",
      "\n",
      "Loading required package: statmod\n",
      "\n",
      "Loading required package: gdsfmt\n",
      "\n",
      "Loading required package: SmartSVA\n",
      "\n",
      "Loading required package: isva\n",
      "\n",
      "Loading required package: qvalue\n",
      "\n",
      "Loading required package: JADE\n",
      "\n",
      "Loading required package: RSpectra\n",
      "\n",
      "Loading required package: preprocessCore\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise, summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## load libraries\n",
    "library(stringr)\n",
    "library(data.table) \n",
    "library(vroom)\n",
    "library(ggplot2)\n",
    "#library(tidyr)\n",
    "#library(limma)\n",
    "library(meffil)\n",
    "library(readxl)\n",
    "library(dplyr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "08604216-f69a-41a0-9a98-dd90d82f77df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set wd\n",
    "setwd('/exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/EBB_methylation/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ab00a641-f210-4be5-8dea-564a1d36c445",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate sample sheet\n",
    "samplesheet <- meffil.create.samplesheet('BrainSamples/data//idats_140716', recursive=TRUE)\n",
    "# update sex and Sample name\n",
    "batch <- fread('metadata/pheno.csv')\n",
    "\n",
    "upd <- \n",
    "fread('BrainSamples/data/idats_140716/Samples_Table_140716.csv') %>% \n",
    "    mutate(Sample_Name = paste0(`Sentrix Barcode`, \"_\", `Array`)) %>%\n",
    "    dplyr::select(c('Sample ID', 'Sample_Name')) %>%\n",
    "    dplyr::rename('Sample_Name2' = 'Sample ID')\n",
    "\n",
    "samplesheet <-\n",
    "inner_join(samplesheet, upd, by = 'Sample_Name') %>% \n",
    "    dplyr::mutate(Sample_Name = Sample_Name2) %>%\n",
    "    dplyr::select(-c(Sample_Name2)) %>%\n",
    "    dplyr::select(-c(Sex)) %>%\n",
    "    left_join(., batch, by = c('Sample_Name'='sample.ID')) %>%\n",
    "    dplyr::rename('Sex' = 'sex')\n",
    "samplesheet <- samplesheet %>% dplyr::select(Sample_Name, Slide, sentrix_row) %>% dplyr::rename(sample.ID = Sample_Name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "63654e4d-513f-4e3f-a703-b868a3cfd78b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## load files\n",
    "# metadata\n",
    "pheno  <- fread('metadata/pheno.csv')\n",
    "# genetic PCs\n",
    "genPCs <- fread('genotypingdata/plink_files/rel_mat/non_imputed_postqc_pca.eigenvec')\n",
    "genPCs <- genPCs[,c('IID','PC1','PC2','PC3')] %>% dplyr::rename(sample.ID = IID)\n",
    "# methylation PCs\n",
    "load('meffil_data/pcs.norm.beta.Robj')\n",
    "# predicted cell counts\n",
    "cellcounts <- fread('meffil_data/cellcounts.txt') %>% dplyr::select(1,3) %>% dplyr::rename(sample.ID = IID, neuron_pct = NeuN_pos)\n",
    "# normalised methulation betas (phenotypes)\n",
    "load('meffil_data/norm.beta.pc10clean.Robj')\n",
    "# outliers \n",
    "outliers <- fread('genotypingdata/plink_files/rel_mat/outliers.txt', header=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd92f7c5-024a-4df7-950a-51d5edd0338c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# INT normalised methylation betas (int phenotypes)\n",
    "load('meffil_data/norm.beta.pc10clean.int.Robj')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86764489-f2de-4453-bb0d-87dc95f450df",
   "metadata": {},
   "source": [
    "### Make covariates file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6fa78be9-9518-4102-bd26-8a795d40afa0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>125</li><li>23</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 125\n",
       "\\item 23\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 125\n",
       "2. 23\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 125  23"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 23</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sample.ID</th><th scope=col>sex</th><th scope=col>age</th><th scope=col>batch</th><th scope=col>tissue</th><th scope=col>PH</th><th scope=col>PMI</th><th scope=col>mPC1</th><th scope=col>mPC2</th><th scope=col>mPC3</th><th scope=col>⋯</th><th scope=col>mPC7</th><th scope=col>mPC8</th><th scope=col>mPC9</th><th scope=col>mPC10</th><th scope=col>PC1</th><th scope=col>PC2</th><th scope=col>PC3</th><th scope=col>neuron_pct</th><th scope=col>Slide</th><th scope=col>sentrix_row</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SD001/07B</td><td>F</td><td>37</td><td>2056G</td><td>Cortex</td><td>6.2</td><td>46</td><td>-13.3000066</td><td> 1.4255424</td><td>-1.12126079</td><td>⋯</td><td> 2.0021279</td><td>-0.02209755</td><td>-0.95143305</td><td> 0.8710113</td><td>-0.00163291</td><td>-0.00233489</td><td> 0.1508320</td><td>0.4903020</td><td>200517480013</td><td>04</td></tr>\n",
       "\t<tr><td>SD001/11B</td><td>M</td><td>74</td><td>2056G</td><td>Cortex</td><td>6.3</td><td>46</td><td> -8.5210675</td><td> 1.8009724</td><td>-0.07410071</td><td>⋯</td><td>-0.4915017</td><td> 0.82199114</td><td>-3.42301636</td><td> 0.9957808</td><td>-0.02005020</td><td> 0.07746340</td><td> 0.1708710</td><td>0.4625232</td><td>200514040135</td><td>01</td></tr>\n",
       "\t<tr><td>SD002/08 </td><td>F</td><td>45</td><td>2056G</td><td>Cortex</td><td>6.4</td><td>28</td><td> -0.1790554</td><td> 0.6709539</td><td> 0.66862374</td><td>⋯</td><td> 0.1442045</td><td>-0.39943974</td><td> 1.49273289</td><td> 1.7907369</td><td> 0.00118622</td><td>-0.01520450</td><td> 0.0279962</td><td>0.4303887</td><td>200516380179</td><td>04</td></tr>\n",
       "\t<tr><td>SD002/10 </td><td>M</td><td>38</td><td>2056G</td><td>Cortex</td><td>6.3</td><td>49</td><td> -6.0827554</td><td> 1.9738718</td><td>-0.44604116</td><td>⋯</td><td> 0.7053378</td><td>-0.33861599</td><td> 2.17009596</td><td>-2.0071594</td><td>-0.04000120</td><td> 0.08611290</td><td>-0.0527741</td><td>0.4407460</td><td>200517480184</td><td>04</td></tr>\n",
       "\t<tr><td>SD003/08 </td><td>M</td><td>50</td><td>2056G</td><td>Cortex</td><td>6.6</td><td>44</td><td>-12.0021204</td><td> 0.1025574</td><td>-0.28077846</td><td>⋯</td><td>-0.6867670</td><td>-0.03828485</td><td>-0.07918727</td><td> 1.0709964</td><td> 0.04792080</td><td> 0.02484220</td><td>-0.0246085</td><td>0.5737720</td><td>200517480069</td><td>07</td></tr>\n",
       "\t<tr><td>SD003/10 </td><td>M</td><td>65</td><td>2056G</td><td>Cortex</td><td>5.6</td><td>34</td><td> -7.8046148</td><td>-5.5463676</td><td>-6.15876494</td><td>⋯</td><td>-1.8515187</td><td>-3.74747226</td><td>-1.66503473</td><td> 1.5856218</td><td>-0.03804790</td><td>-0.07880080</td><td>-0.0421037</td><td>0.4271456</td><td>200526210110</td><td>01</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 23\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " sample.ID & sex & age & batch & tissue & PH & PMI & mPC1 & mPC2 & mPC3 & ⋯ & mPC7 & mPC8 & mPC9 & mPC10 & PC1 & PC2 & PC3 & neuron\\_pct & Slide & sentrix\\_row\\\\\n",
       " <chr> & <chr> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t SD001/07B & F & 37 & 2056G & Cortex & 6.2 & 46 & -13.3000066 &  1.4255424 & -1.12126079 & ⋯ &  2.0021279 & -0.02209755 & -0.95143305 &  0.8710113 & -0.00163291 & -0.00233489 &  0.1508320 & 0.4903020 & 200517480013 & 04\\\\\n",
       "\t SD001/11B & M & 74 & 2056G & Cortex & 6.3 & 46 &  -8.5210675 &  1.8009724 & -0.07410071 & ⋯ & -0.4915017 &  0.82199114 & -3.42301636 &  0.9957808 & -0.02005020 &  0.07746340 &  0.1708710 & 0.4625232 & 200514040135 & 01\\\\\n",
       "\t SD002/08  & F & 45 & 2056G & Cortex & 6.4 & 28 &  -0.1790554 &  0.6709539 &  0.66862374 & ⋯ &  0.1442045 & -0.39943974 &  1.49273289 &  1.7907369 &  0.00118622 & -0.01520450 &  0.0279962 & 0.4303887 & 200516380179 & 04\\\\\n",
       "\t SD002/10  & M & 38 & 2056G & Cortex & 6.3 & 49 &  -6.0827554 &  1.9738718 & -0.44604116 & ⋯ &  0.7053378 & -0.33861599 &  2.17009596 & -2.0071594 & -0.04000120 &  0.08611290 & -0.0527741 & 0.4407460 & 200517480184 & 04\\\\\n",
       "\t SD003/08  & M & 50 & 2056G & Cortex & 6.6 & 44 & -12.0021204 &  0.1025574 & -0.28077846 & ⋯ & -0.6867670 & -0.03828485 & -0.07918727 &  1.0709964 &  0.04792080 &  0.02484220 & -0.0246085 & 0.5737720 & 200517480069 & 07\\\\\n",
       "\t SD003/10  & M & 65 & 2056G & Cortex & 5.6 & 34 &  -7.8046148 & -5.5463676 & -6.15876494 & ⋯ & -1.8515187 & -3.74747226 & -1.66503473 &  1.5856218 & -0.03804790 & -0.07880080 & -0.0421037 & 0.4271456 & 200526210110 & 01\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 23\n",
       "\n",
       "| sample.ID &lt;chr&gt; | sex &lt;chr&gt; | age &lt;int&gt; | batch &lt;chr&gt; | tissue &lt;chr&gt; | PH &lt;dbl&gt; | PMI &lt;dbl&gt; | mPC1 &lt;dbl&gt; | mPC2 &lt;dbl&gt; | mPC3 &lt;dbl&gt; | ⋯ ⋯ | mPC7 &lt;dbl&gt; | mPC8 &lt;dbl&gt; | mPC9 &lt;dbl&gt; | mPC10 &lt;dbl&gt; | PC1 &lt;dbl&gt; | PC2 &lt;dbl&gt; | PC3 &lt;dbl&gt; | neuron_pct &lt;dbl&gt; | Slide &lt;chr&gt; | sentrix_row &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| SD001/07B | F | 37 | 2056G | Cortex | 6.2 | 46 | -13.3000066 |  1.4255424 | -1.12126079 | ⋯ |  2.0021279 | -0.02209755 | -0.95143305 |  0.8710113 | -0.00163291 | -0.00233489 |  0.1508320 | 0.4903020 | 200517480013 | 04 |\n",
       "| SD001/11B | M | 74 | 2056G | Cortex | 6.3 | 46 |  -8.5210675 |  1.8009724 | -0.07410071 | ⋯ | -0.4915017 |  0.82199114 | -3.42301636 |  0.9957808 | -0.02005020 |  0.07746340 |  0.1708710 | 0.4625232 | 200514040135 | 01 |\n",
       "| SD002/08  | F | 45 | 2056G | Cortex | 6.4 | 28 |  -0.1790554 |  0.6709539 |  0.66862374 | ⋯ |  0.1442045 | -0.39943974 |  1.49273289 |  1.7907369 |  0.00118622 | -0.01520450 |  0.0279962 | 0.4303887 | 200516380179 | 04 |\n",
       "| SD002/10  | M | 38 | 2056G | Cortex | 6.3 | 49 |  -6.0827554 |  1.9738718 | -0.44604116 | ⋯ |  0.7053378 | -0.33861599 |  2.17009596 | -2.0071594 | -0.04000120 |  0.08611290 | -0.0527741 | 0.4407460 | 200517480184 | 04 |\n",
       "| SD003/08  | M | 50 | 2056G | Cortex | 6.6 | 44 | -12.0021204 |  0.1025574 | -0.28077846 | ⋯ | -0.6867670 | -0.03828485 | -0.07918727 |  1.0709964 |  0.04792080 |  0.02484220 | -0.0246085 | 0.5737720 | 200517480069 | 07 |\n",
       "| SD003/10  | M | 65 | 2056G | Cortex | 5.6 | 34 |  -7.8046148 | -5.5463676 | -6.15876494 | ⋯ | -1.8515187 | -3.74747226 | -1.66503473 |  1.5856218 | -0.03804790 | -0.07880080 | -0.0421037 | 0.4271456 | 200526210110 | 01 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample.ID sex age batch tissue PH  PMI mPC1        mPC2       mPC3        ⋯ mPC7       mPC8        mPC9        mPC10      PC1         PC2         PC3        neuron_pct Slide        sentrix_row\n",
       "1 SD001/07B F   37  2056G Cortex 6.2 46  -13.3000066  1.4255424 -1.12126079 ⋯  2.0021279 -0.02209755 -0.95143305  0.8710113 -0.00163291 -0.00233489  0.1508320 0.4903020  200517480013 04         \n",
       "2 SD001/11B M   74  2056G Cortex 6.3 46   -8.5210675  1.8009724 -0.07410071 ⋯ -0.4915017  0.82199114 -3.42301636  0.9957808 -0.02005020  0.07746340  0.1708710 0.4625232  200514040135 01         \n",
       "3 SD002/08  F   45  2056G Cortex 6.4 28   -0.1790554  0.6709539  0.66862374 ⋯  0.1442045 -0.39943974  1.49273289  1.7907369  0.00118622 -0.01520450  0.0279962 0.4303887  200516380179 04         \n",
       "4 SD002/10  M   38  2056G Cortex 6.3 49   -6.0827554  1.9738718 -0.44604116 ⋯  0.7053378 -0.33861599  2.17009596 -2.0071594 -0.04000120  0.08611290 -0.0527741 0.4407460  200517480184 04         \n",
       "5 SD003/08  M   50  2056G Cortex 6.6 44  -12.0021204  0.1025574 -0.28077846 ⋯ -0.6867670 -0.03828485 -0.07918727  1.0709964  0.04792080  0.02484220 -0.0246085 0.5737720  200517480069 07         \n",
       "6 SD003/10  M   65  2056G Cortex 5.6 34   -7.8046148 -5.5463676 -6.15876494 ⋯ -1.8515187 -3.74747226 -1.66503473  1.5856218 -0.03804790 -0.07880080 -0.0421037 0.4271456  200526210110 01         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cols_toimpute = c('PMI', 'PH')\n",
    "covariates <- \n",
    "pheno %>% \n",
    "    dplyr::select(sample.ID, sex, age, batch, tissue.region, PH, PMI) %>%\n",
    "    # mean impute PMI and PH \n",
    "    dplyr::mutate(across(all_of(cols_toimpute), ~ ifelse(is.na(.), mean(., na.rm = T), .)),\n",
    "                  PH = round(PH,1),\n",
    "                  PMI = round(PMI,0)\n",
    "                 ) %>%\n",
    "    # add methylation PCs\n",
    "    inner_join(., as_tibble(pcs[,1:10], rownames = 'sample.ID'), by = 'sample.ID') %>%\n",
    "    rename_with(.,  ~ paste0(\"m\", .x ), starts_with('PC')) %>%\n",
    "    # add genetic PCs\n",
    "    inner_join(., genPCs, by = 'sample.ID') %>%\n",
    "    # add cell count\n",
    "    inner_join(., cellcounts, by = 'sample.ID') %>%\n",
    "    # add slide + row\n",
    "    inner_join(., samplesheet,  by = 'sample.ID') %>%\n",
    "    dplyr::rename(tissue = tissue.region) %>%\n",
    "    # exclude outlier\n",
    "    dplyr::filter(!(sample.ID %in% outliers$V1))\n",
    "\n",
    "covariates %>% dim\n",
    "covariates %>% head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc942d6d-9773-48e7-9017-c8fdb18f60c6",
   "metadata": {},
   "source": [
    "### Make phenotypes file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12c8142a-2476-411b-9583-0a761f156977",
   "metadata": {},
   "source": [
    "Phenotypes must be provided in BED format, with a single header line starting with `#` and the first four columns corresponding to: `chr`, `start`, `end`, `phenotype_id`, with **the remaining columns corresponding to samples (the identifiers must match those in the genotype input)**. In addition to .bed/.bed.gz, BED input in .parquet is also supported. **The BED file can specify the center of the cis-window (usually the TSS), with start == end-1,** or alternatively, start and end positions, in which case the cis-window is [start-window, end+window]. A function for generating a BED template from a gene annotation in GTF format is available in pyqtl (io.gtf_to_tss_bed).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3402267d-da35-42ed-9e2e-32ef7eff9ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## load EPIC array annotations\n",
    "anno <- meffil.get.features(\"epic\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "23b77c05-4d46-49e7-8b87-555313d61bf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>815429</li><li>4</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 815429\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 815429\n",
       "2. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 815429      4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>#chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>10524</td><td>10525</td><td>cg14817997</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td>10847</td><td>10848</td><td>cg26928153</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td>10849</td><td>10850</td><td>cg16269199</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td>15864</td><td>15865</td><td>cg13869341</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>18826</td><td>18827</td><td>cg14008030</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>29406</td><td>29407</td><td>cg12045430</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & \\#chr & start & end & name\\\\\n",
       "  & <chr> & <dbl> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 & 10524 & 10525 & cg14817997\\\\\n",
       "\t2 & chr1 & 10847 & 10848 & cg26928153\\\\\n",
       "\t3 & chr1 & 10849 & 10850 & cg16269199\\\\\n",
       "\t4 & chr1 & 15864 & 15865 & cg13869341\\\\\n",
       "\t5 & chr1 & 18826 & 18827 & cg14008030\\\\\n",
       "\t6 & chr1 & 29406 & 29407 & cg12045430\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| <!--/--> | #chr &lt;chr&gt; | start &lt;dbl&gt; | end &lt;int&gt; | name &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | chr1 | 10524 | 10525 | cg14817997 |\n",
       "| 2 | chr1 | 10847 | 10848 | cg26928153 |\n",
       "| 3 | chr1 | 10849 | 10850 | cg16269199 |\n",
       "| 4 | chr1 | 15864 | 15865 | cg13869341 |\n",
       "| 5 | chr1 | 18826 | 18827 | cg14008030 |\n",
       "| 6 | chr1 | 29406 | 29407 | cg12045430 |\n",
       "\n"
      ],
      "text/plain": [
       "  #chr start end   name      \n",
       "1 chr1 10524 10525 cg14817997\n",
       "2 chr1 10847 10848 cg26928153\n",
       "3 chr1 10849 10850 cg16269199\n",
       "4 chr1 15864 15865 cg13869341\n",
       "5 chr1 18826 18827 cg14008030\n",
       "6 chr1 29406 29407 cg12045430"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## make non-sample columns in pheno file\n",
    "pheno0 <-\n",
    "anno %>% \n",
    "    # extract CpGs in the final set\n",
    "    dplyr::filter(name %in% rownames(norm.beta)) %>%\n",
    "    # make BED columns\n",
    "    dplyr::mutate(start = position-1,\n",
    "                  end = position\n",
    "                ) %>%\n",
    "    dplyr::select(chromosome, start, end, name) %>%\n",
    "    dplyr::arrange(chromosome, start) %>%\n",
    "    dplyr::rename(`#chr` = chromosome)\n",
    "pheno0 %>% dim\n",
    "pheno0 %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "be74c7c8-0ade-4f03-8991-c101f89ea03d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>815429</li><li>129</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 815429\n",
       "\\item 129\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 815429\n",
       "2. 129\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 815429    129"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 129</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>#chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>name</th><th scope=col>SD001/11B</th><th scope=col>SD033/10</th><th scope=col>SD024/08</th><th scope=col>SD039/08</th><th scope=col>SD043/06</th><th scope=col>SD034/09B</th><th scope=col>⋯</th><th scope=col>SD024/14B</th><th scope=col>SD008/09</th><th scope=col>SD032/09</th><th scope=col>SD022/08B</th><th scope=col>SD048/12</th><th scope=col>SD055/12</th><th scope=col>SD036/12</th><th scope=col>SD033/08</th><th scope=col>SD025/08</th><th scope=col>SD031/09</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>10524</td><td>10525</td><td>cg14817997</td><td>0.74559588</td><td>0.66332719</td><td>0.6949236</td><td>0.72260903</td><td>0.77141698</td><td>0.74428241</td><td>⋯</td><td>0.66679491</td><td>0.62225889</td><td>0.8216309</td><td>0.8768243</td><td>0.73216765</td><td>0.7234555</td><td>0.75760639</td><td>0.7769564</td><td>0.63724502</td><td>0.77362528</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td>10847</td><td>10848</td><td>cg26928153</td><td>0.94852460</td><td>0.92859978</td><td>0.9085499</td><td>0.94093459</td><td>0.94405633</td><td>0.94615710</td><td>⋯</td><td>0.91189861</td><td>0.91422775</td><td>0.9558368</td><td>0.8865303</td><td>0.93122150</td><td>0.9416910</td><td>0.92200682</td><td>0.9328377</td><td>0.92902259</td><td>0.94013480</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td>10849</td><td>10850</td><td>cg16269199</td><td>0.79303021</td><td>0.82212663</td><td>0.8061039</td><td>0.83840638</td><td>0.84961050</td><td>0.74005705</td><td>⋯</td><td>0.73374094</td><td>0.73715169</td><td>0.9033672</td><td>0.8834348</td><td>0.80570077</td><td>0.8010075</td><td>0.79546215</td><td>0.8017760</td><td>0.72701138</td><td>0.81520981</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td>15864</td><td>15865</td><td>cg13869341</td><td>0.92789490</td><td>0.84442830</td><td>0.9448825</td><td>0.91092827</td><td>0.86397941</td><td>0.88173154</td><td>⋯</td><td>0.90617723</td><td>0.93249506</td><td>0.8977488</td><td>0.8640927</td><td>0.83956341</td><td>0.8777747</td><td>0.93548212</td><td>0.8774606</td><td>0.89908764</td><td>0.88514396</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>18826</td><td>18827</td><td>cg14008030</td><td>0.80972137</td><td>0.70120498</td><td>0.6986296</td><td>0.66519815</td><td>0.81026355</td><td>0.74316353</td><td>⋯</td><td>0.75700377</td><td>0.74628668</td><td>0.8111690</td><td>0.6873779</td><td>0.74092996</td><td>0.7032087</td><td>0.68787400</td><td>0.6922842</td><td>0.77446285</td><td>0.77448368</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>29406</td><td>29407</td><td>cg12045430</td><td>0.08943202</td><td>0.08826019</td><td>0.0825493</td><td>0.07737059</td><td>0.09151521</td><td>0.07083939</td><td>⋯</td><td>0.08673516</td><td>0.07859099</td><td>0.1687048</td><td>0.1036940</td><td>0.09376525</td><td>0.1080019</td><td>0.08736219</td><td>0.0962084</td><td>0.09641446</td><td>0.09894068</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 129\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & \\#chr & start & end & name & SD001/11B & SD033/10 & SD024/08 & SD039/08 & SD043/06 & SD034/09B & ⋯ & SD024/14B & SD008/09 & SD032/09 & SD022/08B & SD048/12 & SD055/12 & SD036/12 & SD033/08 & SD025/08 & SD031/09\\\\\n",
       "  & <chr> & <dbl> & <int> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 & 10524 & 10525 & cg14817997 & 0.74559588 & 0.66332719 & 0.6949236 & 0.72260903 & 0.77141698 & 0.74428241 & ⋯ & 0.66679491 & 0.62225889 & 0.8216309 & 0.8768243 & 0.73216765 & 0.7234555 & 0.75760639 & 0.7769564 & 0.63724502 & 0.77362528\\\\\n",
       "\t2 & chr1 & 10847 & 10848 & cg26928153 & 0.94852460 & 0.92859978 & 0.9085499 & 0.94093459 & 0.94405633 & 0.94615710 & ⋯ & 0.91189861 & 0.91422775 & 0.9558368 & 0.8865303 & 0.93122150 & 0.9416910 & 0.92200682 & 0.9328377 & 0.92902259 & 0.94013480\\\\\n",
       "\t3 & chr1 & 10849 & 10850 & cg16269199 & 0.79303021 & 0.82212663 & 0.8061039 & 0.83840638 & 0.84961050 & 0.74005705 & ⋯ & 0.73374094 & 0.73715169 & 0.9033672 & 0.8834348 & 0.80570077 & 0.8010075 & 0.79546215 & 0.8017760 & 0.72701138 & 0.81520981\\\\\n",
       "\t4 & chr1 & 15864 & 15865 & cg13869341 & 0.92789490 & 0.84442830 & 0.9448825 & 0.91092827 & 0.86397941 & 0.88173154 & ⋯ & 0.90617723 & 0.93249506 & 0.8977488 & 0.8640927 & 0.83956341 & 0.8777747 & 0.93548212 & 0.8774606 & 0.89908764 & 0.88514396\\\\\n",
       "\t5 & chr1 & 18826 & 18827 & cg14008030 & 0.80972137 & 0.70120498 & 0.6986296 & 0.66519815 & 0.81026355 & 0.74316353 & ⋯ & 0.75700377 & 0.74628668 & 0.8111690 & 0.6873779 & 0.74092996 & 0.7032087 & 0.68787400 & 0.6922842 & 0.77446285 & 0.77448368\\\\\n",
       "\t6 & chr1 & 29406 & 29407 & cg12045430 & 0.08943202 & 0.08826019 & 0.0825493 & 0.07737059 & 0.09151521 & 0.07083939 & ⋯ & 0.08673516 & 0.07859099 & 0.1687048 & 0.1036940 & 0.09376525 & 0.1080019 & 0.08736219 & 0.0962084 & 0.09641446 & 0.09894068\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 129\n",
       "\n",
       "| <!--/--> | #chr &lt;chr&gt; | start &lt;dbl&gt; | end &lt;int&gt; | name &lt;chr&gt; | SD001/11B &lt;dbl&gt; | SD033/10 &lt;dbl&gt; | SD024/08 &lt;dbl&gt; | SD039/08 &lt;dbl&gt; | SD043/06 &lt;dbl&gt; | SD034/09B &lt;dbl&gt; | ⋯ ⋯ | SD024/14B &lt;dbl&gt; | SD008/09 &lt;dbl&gt; | SD032/09 &lt;dbl&gt; | SD022/08B &lt;dbl&gt; | SD048/12 &lt;dbl&gt; | SD055/12 &lt;dbl&gt; | SD036/12 &lt;dbl&gt; | SD033/08 &lt;dbl&gt; | SD025/08 &lt;dbl&gt; | SD031/09 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 | 10524 | 10525 | cg14817997 | 0.74559588 | 0.66332719 | 0.6949236 | 0.72260903 | 0.77141698 | 0.74428241 | ⋯ | 0.66679491 | 0.62225889 | 0.8216309 | 0.8768243 | 0.73216765 | 0.7234555 | 0.75760639 | 0.7769564 | 0.63724502 | 0.77362528 |\n",
       "| 2 | chr1 | 10847 | 10848 | cg26928153 | 0.94852460 | 0.92859978 | 0.9085499 | 0.94093459 | 0.94405633 | 0.94615710 | ⋯ | 0.91189861 | 0.91422775 | 0.9558368 | 0.8865303 | 0.93122150 | 0.9416910 | 0.92200682 | 0.9328377 | 0.92902259 | 0.94013480 |\n",
       "| 3 | chr1 | 10849 | 10850 | cg16269199 | 0.79303021 | 0.82212663 | 0.8061039 | 0.83840638 | 0.84961050 | 0.74005705 | ⋯ | 0.73374094 | 0.73715169 | 0.9033672 | 0.8834348 | 0.80570077 | 0.8010075 | 0.79546215 | 0.8017760 | 0.72701138 | 0.81520981 |\n",
       "| 4 | chr1 | 15864 | 15865 | cg13869341 | 0.92789490 | 0.84442830 | 0.9448825 | 0.91092827 | 0.86397941 | 0.88173154 | ⋯ | 0.90617723 | 0.93249506 | 0.8977488 | 0.8640927 | 0.83956341 | 0.8777747 | 0.93548212 | 0.8774606 | 0.89908764 | 0.88514396 |\n",
       "| 5 | chr1 | 18826 | 18827 | cg14008030 | 0.80972137 | 0.70120498 | 0.6986296 | 0.66519815 | 0.81026355 | 0.74316353 | ⋯ | 0.75700377 | 0.74628668 | 0.8111690 | 0.6873779 | 0.74092996 | 0.7032087 | 0.68787400 | 0.6922842 | 0.77446285 | 0.77448368 |\n",
       "| 6 | chr1 | 29406 | 29407 | cg12045430 | 0.08943202 | 0.08826019 | 0.0825493 | 0.07737059 | 0.09151521 | 0.07083939 | ⋯ | 0.08673516 | 0.07859099 | 0.1687048 | 0.1036940 | 0.09376525 | 0.1080019 | 0.08736219 | 0.0962084 | 0.09641446 | 0.09894068 |\n",
       "\n"
      ],
      "text/plain": [
       "  #chr start end   name       SD001/11B  SD033/10   SD024/08  SD039/08   SD043/06   SD034/09B  ⋯ SD024/14B  SD008/09   SD032/09  SD022/08B SD048/12   SD055/12  SD036/12   SD033/08  SD025/08  \n",
       "1 chr1 10524 10525 cg14817997 0.74559588 0.66332719 0.6949236 0.72260903 0.77141698 0.74428241 ⋯ 0.66679491 0.62225889 0.8216309 0.8768243 0.73216765 0.7234555 0.75760639 0.7769564 0.63724502\n",
       "2 chr1 10847 10848 cg26928153 0.94852460 0.92859978 0.9085499 0.94093459 0.94405633 0.94615710 ⋯ 0.91189861 0.91422775 0.9558368 0.8865303 0.93122150 0.9416910 0.92200682 0.9328377 0.92902259\n",
       "3 chr1 10849 10850 cg16269199 0.79303021 0.82212663 0.8061039 0.83840638 0.84961050 0.74005705 ⋯ 0.73374094 0.73715169 0.9033672 0.8834348 0.80570077 0.8010075 0.79546215 0.8017760 0.72701138\n",
       "4 chr1 15864 15865 cg13869341 0.92789490 0.84442830 0.9448825 0.91092827 0.86397941 0.88173154 ⋯ 0.90617723 0.93249506 0.8977488 0.8640927 0.83956341 0.8777747 0.93548212 0.8774606 0.89908764\n",
       "5 chr1 18826 18827 cg14008030 0.80972137 0.70120498 0.6986296 0.66519815 0.81026355 0.74316353 ⋯ 0.75700377 0.74628668 0.8111690 0.6873779 0.74092996 0.7032087 0.68787400 0.6922842 0.77446285\n",
       "6 chr1 29406 29407 cg12045430 0.08943202 0.08826019 0.0825493 0.07737059 0.09151521 0.07083939 ⋯ 0.08673516 0.07859099 0.1687048 0.1036940 0.09376525 0.1080019 0.08736219 0.0962084 0.09641446\n",
       "  SD031/09  \n",
       "1 0.77362528\n",
       "2 0.94013480\n",
       "3 0.81520981\n",
       "4 0.88514396\n",
       "5 0.77448368\n",
       "6 0.09894068"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "phenotypes <- \n",
    "    inner_join(pheno0, as_tibble(norm.beta, rownames = 'name'), by = 'name') %>%\n",
    "    # remove outliers\n",
    "    select(-c(outliers$V1))\n",
    "dim(phenotypes)\n",
    "head(phenotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "66f3f3ee-dc41-4882-b7e9-d4b33957e427",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>815429</li><li>129</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 815429\n",
       "\\item 129\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 815429\n",
       "2. 129\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 815429    129"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 129</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>#chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>name</th><th scope=col>SD001/11B</th><th scope=col>SD033/10</th><th scope=col>SD024/08</th><th scope=col>SD039/08</th><th scope=col>SD043/06</th><th scope=col>SD034/09B</th><th scope=col>⋯</th><th scope=col>SD024/14B</th><th scope=col>SD008/09</th><th scope=col>SD032/09</th><th scope=col>SD022/08B</th><th scope=col>SD048/12</th><th scope=col>SD055/12</th><th scope=col>SD036/12</th><th scope=col>SD033/08</th><th scope=col>SD025/08</th><th scope=col>SD031/09</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>10524</td><td>10525</td><td>cg14817997</td><td> 0.57766241</td><td>-0.5543124</td><td>-0.1699120</td><td> 0.06968492</td><td> 1.0501464</td><td> 0.53126076</td><td>⋯</td><td>-0.5312608</td><td>-0.833146866</td><td> 1.914506</td><td> 2.6547590</td><td> 0.29258253</td><td> 0.08964235</td><td> 0.80532021</td><td> 1.12190047</td><td>-0.7514515</td><td> 1.085324908</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td>10847</td><td>10848</td><td>cg26928153</td><td> 1.43673379</td><td>-0.3134112</td><td>-1.4949763</td><td> 0.72532289</td><td> 0.9208230</td><td> 1.08532491</td><td>⋯</td><td>-1.3329760</td><td>-1.160041903</td><td> 2.057038</td><td>-2.2601890</td><td>-0.08964235</td><td> 0.77810361</td><td>-0.95165921</td><td> 0.04975521</td><td>-0.2512934</td><td> 0.625342896</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td>10849</td><td>10850</td><td>cg16269199</td><td> 0.06968492</td><td> 0.6996803</td><td> 0.3767637</td><td> 0.86163412</td><td> 1.2418668</td><td>-0.95165921</td><td>⋯</td><td>-1.1600419</td><td>-1.121900467</td><td> 2.654759</td><td> 1.9145058</td><td> 0.33437681</td><td> 0.21042839</td><td> 0.08964235</td><td> 0.23081270</td><td>-1.2418668</td><td> 0.625342896</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td>15864</td><td>15865</td><td>cg13869341</td><td> 0.72532289</td><td>-1.4367338</td><td> 1.3829941</td><td> 0.33437681</td><td>-1.0501464</td><td>-0.53126076</td><td>⋯</td><td> 0.1096356</td><td> 0.920822976</td><td>-0.169912</td><td>-1.0162217</td><td>-1.55878355</td><td>-0.55431238</td><td> 1.01622173</td><td>-0.60133175</td><td>-0.1497620</td><td>-0.485976078</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>18826</td><td>18827</td><td>cg14008030</td><td> 1.01622173</td><td>-0.7253229</td><td>-0.7514515</td><td>-1.62965420</td><td> 1.0501464</td><td>-0.06968492</td><td>⋯</td><td> 0.2308127</td><td> 0.009947102</td><td> 1.085325</td><td>-1.0853249</td><td>-0.12967268</td><td>-0.67448975</td><td>-1.05014643</td><td>-0.86163412</td><td> 0.3767637</td><td> 0.398208921</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>29406</td><td>29407</td><td>cg12045430</td><td>-0.50848806</td><td>-0.5776624</td><td>-1.0162217</td><td>-1.49497627</td><td>-0.3343768</td><td>-1.80274309</td><td>⋯</td><td>-0.6996803</td><td>-1.382994127</td><td> 2.654759</td><td> 0.2925825</td><td>-0.27188001</td><td> 0.57766241</td><td>-0.62534290</td><td>-0.16991197</td><td>-0.1497620</td><td> 0.009947102</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 129\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & \\#chr & start & end & name & SD001/11B & SD033/10 & SD024/08 & SD039/08 & SD043/06 & SD034/09B & ⋯ & SD024/14B & SD008/09 & SD032/09 & SD022/08B & SD048/12 & SD055/12 & SD036/12 & SD033/08 & SD025/08 & SD031/09\\\\\n",
       "  & <chr> & <dbl> & <int> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 & 10524 & 10525 & cg14817997 &  0.57766241 & -0.5543124 & -0.1699120 &  0.06968492 &  1.0501464 &  0.53126076 & ⋯ & -0.5312608 & -0.833146866 &  1.914506 &  2.6547590 &  0.29258253 &  0.08964235 &  0.80532021 &  1.12190047 & -0.7514515 &  1.085324908\\\\\n",
       "\t2 & chr1 & 10847 & 10848 & cg26928153 &  1.43673379 & -0.3134112 & -1.4949763 &  0.72532289 &  0.9208230 &  1.08532491 & ⋯ & -1.3329760 & -1.160041903 &  2.057038 & -2.2601890 & -0.08964235 &  0.77810361 & -0.95165921 &  0.04975521 & -0.2512934 &  0.625342896\\\\\n",
       "\t3 & chr1 & 10849 & 10850 & cg16269199 &  0.06968492 &  0.6996803 &  0.3767637 &  0.86163412 &  1.2418668 & -0.95165921 & ⋯ & -1.1600419 & -1.121900467 &  2.654759 &  1.9145058 &  0.33437681 &  0.21042839 &  0.08964235 &  0.23081270 & -1.2418668 &  0.625342896\\\\\n",
       "\t4 & chr1 & 15864 & 15865 & cg13869341 &  0.72532289 & -1.4367338 &  1.3829941 &  0.33437681 & -1.0501464 & -0.53126076 & ⋯ &  0.1096356 &  0.920822976 & -0.169912 & -1.0162217 & -1.55878355 & -0.55431238 &  1.01622173 & -0.60133175 & -0.1497620 & -0.485976078\\\\\n",
       "\t5 & chr1 & 18826 & 18827 & cg14008030 &  1.01622173 & -0.7253229 & -0.7514515 & -1.62965420 &  1.0501464 & -0.06968492 & ⋯ &  0.2308127 &  0.009947102 &  1.085325 & -1.0853249 & -0.12967268 & -0.67448975 & -1.05014643 & -0.86163412 &  0.3767637 &  0.398208921\\\\\n",
       "\t6 & chr1 & 29406 & 29407 & cg12045430 & -0.50848806 & -0.5776624 & -1.0162217 & -1.49497627 & -0.3343768 & -1.80274309 & ⋯ & -0.6996803 & -1.382994127 &  2.654759 &  0.2925825 & -0.27188001 &  0.57766241 & -0.62534290 & -0.16991197 & -0.1497620 &  0.009947102\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 129\n",
       "\n",
       "| <!--/--> | #chr &lt;chr&gt; | start &lt;dbl&gt; | end &lt;int&gt; | name &lt;chr&gt; | SD001/11B &lt;dbl&gt; | SD033/10 &lt;dbl&gt; | SD024/08 &lt;dbl&gt; | SD039/08 &lt;dbl&gt; | SD043/06 &lt;dbl&gt; | SD034/09B &lt;dbl&gt; | ⋯ ⋯ | SD024/14B &lt;dbl&gt; | SD008/09 &lt;dbl&gt; | SD032/09 &lt;dbl&gt; | SD022/08B &lt;dbl&gt; | SD048/12 &lt;dbl&gt; | SD055/12 &lt;dbl&gt; | SD036/12 &lt;dbl&gt; | SD033/08 &lt;dbl&gt; | SD025/08 &lt;dbl&gt; | SD031/09 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 | 10524 | 10525 | cg14817997 |  0.57766241 | -0.5543124 | -0.1699120 |  0.06968492 |  1.0501464 |  0.53126076 | ⋯ | -0.5312608 | -0.833146866 |  1.914506 |  2.6547590 |  0.29258253 |  0.08964235 |  0.80532021 |  1.12190047 | -0.7514515 |  1.085324908 |\n",
       "| 2 | chr1 | 10847 | 10848 | cg26928153 |  1.43673379 | -0.3134112 | -1.4949763 |  0.72532289 |  0.9208230 |  1.08532491 | ⋯ | -1.3329760 | -1.160041903 |  2.057038 | -2.2601890 | -0.08964235 |  0.77810361 | -0.95165921 |  0.04975521 | -0.2512934 |  0.625342896 |\n",
       "| 3 | chr1 | 10849 | 10850 | cg16269199 |  0.06968492 |  0.6996803 |  0.3767637 |  0.86163412 |  1.2418668 | -0.95165921 | ⋯ | -1.1600419 | -1.121900467 |  2.654759 |  1.9145058 |  0.33437681 |  0.21042839 |  0.08964235 |  0.23081270 | -1.2418668 |  0.625342896 |\n",
       "| 4 | chr1 | 15864 | 15865 | cg13869341 |  0.72532289 | -1.4367338 |  1.3829941 |  0.33437681 | -1.0501464 | -0.53126076 | ⋯ |  0.1096356 |  0.920822976 | -0.169912 | -1.0162217 | -1.55878355 | -0.55431238 |  1.01622173 | -0.60133175 | -0.1497620 | -0.485976078 |\n",
       "| 5 | chr1 | 18826 | 18827 | cg14008030 |  1.01622173 | -0.7253229 | -0.7514515 | -1.62965420 |  1.0501464 | -0.06968492 | ⋯ |  0.2308127 |  0.009947102 |  1.085325 | -1.0853249 | -0.12967268 | -0.67448975 | -1.05014643 | -0.86163412 |  0.3767637 |  0.398208921 |\n",
       "| 6 | chr1 | 29406 | 29407 | cg12045430 | -0.50848806 | -0.5776624 | -1.0162217 | -1.49497627 | -0.3343768 | -1.80274309 | ⋯ | -0.6996803 | -1.382994127 |  2.654759 |  0.2925825 | -0.27188001 |  0.57766241 | -0.62534290 | -0.16991197 | -0.1497620 |  0.009947102 |\n",
       "\n"
      ],
      "text/plain": [
       "  #chr start end   name       SD001/11B   SD033/10   SD024/08   SD039/08    SD043/06   SD034/09B   ⋯ SD024/14B  SD008/09     SD032/09  SD022/08B  SD048/12    SD055/12    SD036/12    SD033/08   \n",
       "1 chr1 10524 10525 cg14817997  0.57766241 -0.5543124 -0.1699120  0.06968492  1.0501464  0.53126076 ⋯ -0.5312608 -0.833146866  1.914506  2.6547590  0.29258253  0.08964235  0.80532021  1.12190047\n",
       "2 chr1 10847 10848 cg26928153  1.43673379 -0.3134112 -1.4949763  0.72532289  0.9208230  1.08532491 ⋯ -1.3329760 -1.160041903  2.057038 -2.2601890 -0.08964235  0.77810361 -0.95165921  0.04975521\n",
       "3 chr1 10849 10850 cg16269199  0.06968492  0.6996803  0.3767637  0.86163412  1.2418668 -0.95165921 ⋯ -1.1600419 -1.121900467  2.654759  1.9145058  0.33437681  0.21042839  0.08964235  0.23081270\n",
       "4 chr1 15864 15865 cg13869341  0.72532289 -1.4367338  1.3829941  0.33437681 -1.0501464 -0.53126076 ⋯  0.1096356  0.920822976 -0.169912 -1.0162217 -1.55878355 -0.55431238  1.01622173 -0.60133175\n",
       "5 chr1 18826 18827 cg14008030  1.01622173 -0.7253229 -0.7514515 -1.62965420  1.0501464 -0.06968492 ⋯  0.2308127  0.009947102  1.085325 -1.0853249 -0.12967268 -0.67448975 -1.05014643 -0.86163412\n",
       "6 chr1 29406 29407 cg12045430 -0.50848806 -0.5776624 -1.0162217 -1.49497627 -0.3343768 -1.80274309 ⋯ -0.6996803 -1.382994127  2.654759  0.2925825 -0.27188001  0.57766241 -0.62534290 -0.16991197\n",
       "  SD025/08   SD031/09    \n",
       "1 -0.7514515  1.085324908\n",
       "2 -0.2512934  0.625342896\n",
       "3 -1.2418668  0.625342896\n",
       "4 -0.1497620 -0.485976078\n",
       "5  0.3767637  0.398208921\n",
       "6 -0.1497620  0.009947102"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "phenotypes.int <- \n",
    "    inner_join(pheno0, as_tibble(norm.beta.int, rownames = 'name'), by = 'name') %>%\n",
    "    # remove outliers\n",
    "    select(-c(outliers$V1))\n",
    "dim(phenotypes.int)\n",
    "head(phenotypes.int)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82f3af35-5616-43a5-a2bf-4908633aedaa",
   "metadata": {},
   "source": [
    "### Match sample IDs across genotypes, phenotypes and covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a400245b-8d21-4365-8b10-496552084751",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## match with genotype sample names\n",
    "psam <- fread('genotypingdata/plink_files/pgen/imputed_allchr_newIDs_noOutlier.psam')\n",
    "phenotypes <- phenotypes %>% select(1:4, psam$IID)\n",
    "all(psam$IID == colnames(phenotypes)[5:ncol(phenotypes)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "526d777e-be70-44d0-a4b3-0bfc1bfcd793",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "phenotypes.int <- phenotypes.int %>% select(1:4, psam$IID)\n",
    "all(psam$IID == colnames(phenotypes.int)[5:ncol(phenotypes.int)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "730b1eab-7097-42b4-aa30-37e42159a0a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## match with covariate sample names\n",
    "covariates <- covariates[match(psam$IID, covariates$sample.ID),]\n",
    "all(psam$IID == covariates$sample.ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bbd3db2-9285-474c-afe4-b498f818972d",
   "metadata": {},
   "source": [
    "### Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7583e0fc-1bf4-4d24-a44e-3e57e65149aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## save\n",
    "fwrite(covariates, 'tensorqtl/covariates.txt', sep = \"\\t\")\n",
    "fwrite(phenotypes, 'tensorqtl/phenotypes.bed', sep = \"\\t\")\n",
    "fwrite(phenotypes.int, 'tensorqtl/phenotypes.int.bed', sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "60cf66f6-7b28-4f50-9d3e-6d665848a468",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.5.1 (2025-06-13)\n",
       "Platform: x86_64-conda-linux-gnu\n",
       "Running under: Rocky Linux 9.5 (Blue Onyx)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /exports/cmvm/eddie/smgphs/groups/Quantgen/Users/vasilis/PHD/jupiter-setup/envs/jpt/lib/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8      \n",
       " [8] LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Europe/London\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] dplyr_1.1.4           readxl_1.4.5          meffil_1.5.1          preprocessCore_1.70.0 SmartSVA_0.1.3        RSpectra_0.16-2       isva_1.9              JADE_2.0-4           \n",
       " [9] qvalue_2.40.0         gdsfmt_1.44.1         statmod_1.5.1         quadprog_1.5-8        DNAcopy_1.82.0        fastICA_1.2-7         lme4_1.1-37           Matrix_1.7-4         \n",
       "[17] multcomp_1.4-29       TH.data_1.1-4         survival_3.8-3        mvtnorm_1.3-3         matrixStats_1.5.0     markdown_2.0          gridExtra_2.3         Cairo_1.7-0          \n",
       "[25] knitr_1.50            reshape2_1.4.5        plyr_1.8.9            sva_3.56.0            BiocParallel_1.42.2   genefilter_1.90.0     mgcv_1.9-4            nlme_3.1-168         \n",
       "[33] limma_3.64.3          sandwich_3.1-1        lmtest_0.9-40         zoo_1.8-14            MASS_7.3-65           illuminaio_0.50.0     ggplot2_4.0.1         vroom_1.6.6          \n",
       "[41] data.table_1.17.8     stringr_1.6.0        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] Rdpack_2.6.4            DBI_1.2.3               rlang_1.1.6             magrittr_2.0.4          clue_0.3-66             compiler_4.5.1          RSQLite_2.4.4           png_0.1-8              \n",
       " [9] vctrs_0.6.5             pkgconfig_2.0.3         crayon_1.5.3            fastmap_1.2.0           XVector_0.48.0          tzdb_0.5.0              nloptr_2.2.1            UCSC.utils_1.4.0       \n",
       "[17] bit_4.6.0               xfun_0.54               cachem_1.1.0            GenomeInfoDb_1.44.3     jsonlite_2.0.0          blob_1.2.4              uuid_1.2-1              cluster_2.1.8.1        \n",
       "[25] R6_2.6.1                stringi_1.8.7           RColorBrewer_1.1-3      boot_1.3-32             cellranger_1.1.0        Rcpp_1.1.0              IRkernel_1.3.2          base64enc_0.1-3        \n",
       "[33] IRanges_2.42.0          splines_4.5.1           tidyselect_1.2.1        dichromat_2.0-0.1       codetools_0.2-20        lattice_0.22-7          tibble_3.3.0            Biobase_2.68.0         \n",
       "[41] withr_3.0.2             KEGGREST_1.48.1         S7_0.2.1                askpass_1.2.1           evaluate_1.0.5          Biostrings_2.76.0       pillar_1.11.1           MatrixGenerics_1.20.0  \n",
       "[49] stats4_4.5.1            reformulas_0.4.2        generics_0.1.4          IRdisplay_1.1           S4Vectors_0.46.0        scales_1.4.0            minqa_1.2.8             xtable_1.8-4           \n",
       "[57] glue_1.8.0              tools_4.5.1             annotate_1.86.1         locfit_1.5-9.12         pbdZMQ_0.3-14           XML_3.99-0.19           grid_4.5.1              rbibutils_2.4          \n",
       "[65] AnnotationDbi_1.70.0    edgeR_4.6.3             base64_2.0.2            GenomeInfoDbData_1.2.14 repr_1.1.7              cli_3.6.5               gtable_0.3.6            digest_0.6.38          \n",
       "[73] BiocGenerics_0.54.1     farver_2.1.2            memoise_2.0.1           htmltools_0.5.8.1       lifecycle_1.0.4         httr_1.4.7              openssl_2.3.4           bit64_4.6.0-1          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
