{"metadata": {"language_info": {"name": "python", "version": "3.12.9", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}, "kernelspec": {"name": "python3", "display_name": "Python 3 (ipykernel)", "language": "python"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "This notebook prepares the genomic data for imputation of brain CpG levels from Edinburh Brain Bank (EBB) to UKB genotypes.\n- get bgen variant ids (chr:pos_a1_a2) that correspond to EBB variant ids (chr:pos:a1:a2); note: alleles may be swapped so get chr:pos:a1:a2 and chr:pos:a2:a1 EBB variant ids and find the intersect wiht bgen ids.\n- extract the bgen variant ids (so i dont have to update the large bgen files)\n- Sample and variant QC\n- update the bgen variant ids with the ebb variant ids\n- Date: 03.02.2026\n", "metadata": {"tags": []}}, {"cell_type": "markdown", "source": "## Setup", "metadata": {"tags": []}}, {"cell_type": "code", "source": "%%bash\npip install openpyxl\npip install -U scikit-learn\npip install statsmodels", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": [{"name": "stdout", "text": "Requirement already satisfied: openpyxl in /opt/conda/lib/python3.12/site-packages (3.1.5)\nRequirement already satisfied: et-xmlfile in /opt/conda/lib/python3.12/site-packages (from openpyxl) (2.0.0)\n", "output_type": "stream"}, {"name": "stderr", "text": "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.\u001b[0m\u001b[33m\n", "output_type": "stream"}, {"name": "stdout", "text": "Requirement already satisfied: scikit-learn in /opt/conda/lib/python3.12/site-packages (1.8.0)\nRequirement already satisfied: numpy>=1.24.1 in /opt/conda/lib/python3.12/site-packages (from scikit-learn) (1.26.4)\nRequirement already satisfied: scipy>=1.10.0 in /opt/conda/lib/python3.12/site-packages (from scikit-learn) (1.15.3)\nRequirement already satisfied: joblib>=1.3.0 in /opt/conda/lib/python3.12/site-packages (from scikit-learn) (1.5.3)\nRequirement already satisfied: threadpoolctl>=3.2.0 in /opt/conda/lib/python3.12/site-packages (from scikit-learn) (3.6.0)\n", "output_type": "stream"}, {"name": "stderr", "text": "\u001b[0m\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.\u001b[0m\u001b[33m\n", "output_type": "stream"}, {"name": "stdout", "text": "Requirement already satisfied: statsmodels in /opt/conda/lib/python3.12/site-packages (0.14.6)\nRequirement already satisfied: numpy<3,>=1.22.3 in /opt/conda/lib/python3.12/site-packages (from statsmodels) (1.26.4)\nRequirement already satisfied: scipy!=1.9.2,>=1.8 in /opt/conda/lib/python3.12/site-packages (from statsmodels) (1.15.3)\nRequirement already satisfied: pandas!=2.1.0,>=1.4 in /opt/conda/lib/python3.12/site-packages (from statsmodels) (2.2.3)\nRequirement already satisfied: patsy>=0.5.6 in /opt/conda/lib/python3.12/site-packages (from statsmodels) (1.0.2)\nRequirement already satisfied: packaging>=21.3 in /opt/conda/lib/python3.12/site-packages (from statsmodels) (24.2)\nRequirement already satisfied: python-dateutil>=2.8.2 in /opt/conda/lib/python3.12/site-packages (from pandas!=2.1.0,>=1.4->statsmodels) (2.9.0.post0)\nRequirement already satisfied: pytz>=2020.1 in /opt/conda/lib/python3.12/site-packages (from pandas!=2.1.0,>=1.4->statsmodels) (2025.2)\nRequirement already satisfied: tzdata>=2022.7 in /opt/conda/lib/python3.12/site-packages (from pandas!=2.1.0,>=1.4->statsmodels) (2025.2)\nRequirement already satisfied: six>=1.5 in /opt/conda/lib/python3.12/site-packages (from python-dateutil>=2.8.2->pandas!=2.1.0,>=1.4->statsmodels) (1.17.0)\n", "output_type": "stream"}, {"name": "stderr", "text": "\u001b[0m\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.\u001b[0m\u001b[33m\n\u001b[0m", "output_type": "stream"}]}, {"cell_type": "code", "source": "import os\nimport glob\nimport pandas as pd\nfrom pandas.core.common import flatten\nimport re\nimport numpy as np\nimport seaborn as sns\nfrom sklearn import datasets, linear_model, metrics\nimport statsmodels.api as sm\nimport statsmodels.formula.api as smf", "metadata": {"trusted": true, "tags": []}, "execution_count": 13, "outputs": []}, {"cell_type": "markdown", "source": "### load EBB weights ", "metadata": {"tags": []}}, {"cell_type": "code", "source": "%%bash\ndx download -f vasilis/data/ebb/weights/EBB.BRAIN.METHYL.HERIT.tar.bz2 ", "metadata": {"trusted": true, "tags": []}, "execution_count": 30, "outputs": []}, {"cell_type": "code", "source": "%%bash\ntar -xjf EBB.BRAIN.METHYL.HERIT.tar.bz2 ", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ndx upload -r EBB.BRAIN.METHYL.HERIT --dest vasilis/data/ebb/weights/", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "### load UKB imputation data", "metadata": {"tags": []}}, {"cell_type": "code", "source": "%%bash\nmkdir -p imp/\ndx download -f -o imp/ Bulk/Imputation/'UKB imputation from genotype'/ukb22828_c*_b0_v3.mfi.txt ", "metadata": {"trusted": true, "tags": []}, "execution_count": 5, "outputs": []}, {"cell_type": "markdown", "source": "## Filter samples\n\n- no sex missmatch: p22001 vs p31\n- white british ancestry: p22006\n- no heterozygosiry or missingness outlier: p22027\n- no \\>\\=10 3rd degree relatives in dataset: p22021\n- no sex chromosome aneuploidy: p22019\n", "metadata": {"tags": []}}, {"cell_type": "code", "source": "all = pd.read_csv('/mnt/project/vasilis/data/pheno_all_unfiltered.csv')", "metadata": {"trusted": true, "tags": []}, "execution_count": 50, "outputs": [{"name": "stderr", "text": "/tmp/ipykernel_1656/207872071.py:1: DtypeWarning: Columns (63,64,73,74,75,76,77,78,80,81,82,83,84,85,86,87,88,100) have mixed types. Specify dtype option on import or set low_memory=False.\n  all = pd.read_csv('/mnt/project/vasilis/data/pheno_all_unfiltered.csv')\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "all['p22001'] = pd.to_numeric(all['p22001'], errors=\"coerce\").astype(\"Int64\")\ncols = ['eid', 'p22006', 'p22027', 'p22021', 'p22019', 'p22001', 'p31']\nqc = all[cols]\nqc.head()", "metadata": {"trusted": true, "tags": []}, "execution_count": 94, "outputs": [{"execution_count": 94, "output_type": "execute_result", "data": {"text/plain": "       eid  p22006  p22027  p22021  p22019  p22001  p31\n0  1000015     1.0     NaN     0.0     NaN       1    1\n1  1000053     1.0     NaN     1.0     NaN       0    0\n2  1000132     1.0     NaN     0.0     NaN       0    0\n3  1000148     1.0     NaN     0.0     NaN       0    0\n4  1000163     1.0     NaN     0.0     NaN       1    1", "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>eid</th>\n      <th>p22006</th>\n      <th>p22027</th>\n      <th>p22021</th>\n      <th>p22019</th>\n      <th>p22001</th>\n      <th>p31</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1000015</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>0.0</td>\n      <td>NaN</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1000053</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>1000132</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>0.0</td>\n      <td>NaN</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>1000148</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>0.0</td>\n      <td>NaN</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>1000163</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>0.0</td>\n      <td>NaN</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n  </tbody>\n</table>\n</div>"}, "metadata": {}}]}, {"cell_type": "code", "source": "qced= qc.loc[\n    (qc['p22006'] == 1) & \n    (qc['p22001'] == qc['p31']) &\n    (qc['p22021'] != 10) &\n    (qc['p22027'].isna()) &\n    (qc['p22019'].isna())\n]\nlen(qced['eid'])", "metadata": {"trusted": true}, "execution_count": 113, "outputs": [{"execution_count": 113, "output_type": "execute_result", "data": {"text/plain": "407827"}, "metadata": {}}]}, {"cell_type": "code", "source": "# make .keep file\nqced[['eid', 'eid']].to_csv('wb.eids', index=False, header=False, sep=\" \")\n!dx upload wb.eids --dest vasilis/data/ebb/", "metadata": {"trusted": true, "tags": []}, "execution_count": 131, "outputs": [{"name": "stdout", "text": "[===========================================================>] Uploaded 6,525,232 of 6,525,232 bytes (100%) wb.eids\nID                                file-J60Kx0jJZB71xj86Gg2kFfQ4\nClass                             file\nProject                           project-GfvP6PQJZB72v2Vk348Bb2yg\nFolder                            /vasilis/data/ebb\nName                              wb.eids\nState                             \u001b[33mclosing\u001b[0m\nVisibility                        visible\nTypes                             -\nProperties                        -\nTags                              -\nOutgoing links                    -\nCreated                           Mon Feb  2 18:35:47 2026\nCreated by                        vasilisraptis\n via the job                      job-J607PJ0JZB75g08bgjY771xk\nLast modified                     Mon Feb  2 18:35:48 2026\nMedia type                        \narchivalState                     \"live\"\ncloudAccount                      \"cloudaccount-dnanexus\"\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## Extract weights' variants & QC variants", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n## get EBB variant ids\nawk '{print $1}' EBB.BRAIN.METHYL.HERIT/EBB.BRAIN.METHYL.HERIT.variants | sort | uniq > ebb.variants.1\n## also get the swapped variant ids\ncat ebb.variants.1 | awk 'BEGIN{FS=\":\"} {print $1\":\"$2\":\"$4\":\"$3}' | sort | uniq > ebb.variants.2\nwc -l ebb.variants.1\nwc -l ebb.variants.2", "metadata": {"trusted": true, "tags": []}, "execution_count": 6, "outputs": [{"name": "stdout", "text": "473982 ebb.variants.1\n473982 ebb.variants.2\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "%%bash\nmkdir -p imp/temp\nmkdir -p imp/extract\nfor chr in {1..22}\ndo\n    # filter INFO > 0.8 & MAF > 0.01; create ebb variant format: chr:pos:a1:a2; sort (for joining)\n    awk -v chr=$chr '$NF > 0.8 && $6 > 0.01 { print chr\":\"$3\":\"$4\":\"$5, $2}' imp/ukb22828_c${chr}_b0_v3.mfi.txt | sort | uniq > imp/temp/temp_c${chr}.variants\n    # find ebb variants in ukb (intersect)\n    join -1 1 -2 1 ebb.variants.1 imp/temp/temp_c${chr}.variants > imp/extract/imp_c${chr}.extract1\n    join -1 1 -2 1 ebb.variants.2 imp/temp/temp_c${chr}.variants > imp/extract/imp_c${chr}.extract2\n    # keep both formats\n    cat imp/extract/imp_c${chr}.extract1 imp/extract/imp_c${chr}.extract2 | awk '{print $0}' | sort | uniq > imp/extract/imp_c${chr}.bothIDs.extract\n    # keep only ukb format\n    awk '{print $2}' imp/extract/imp_c${chr}.bothIDs.extract | sort | uniq > imp/extract/imp_c${chr}.extract\n    wc -l imp/extract/imp_c${chr}.extract\n    rm imp/extract/imp_c${chr}.extract1 imp/extract/imp_c${chr}.extract2\ndone", "metadata": {"trusted": true, "tags": []}, "execution_count": 25, "outputs": [{"name": "stdout", "text": "44171 imp/extract/imp_c1.extract\n37344 imp/extract/imp_c2.extract\n28347 imp/extract/imp_c3.extract\n22820 imp/extract/imp_c4.extract\n26461 imp/extract/imp_c5.extract\n29832 imp/extract/imp_c6.extract\n26394 imp/extract/imp_c7.extract\n22743 imp/extract/imp_c8.extract\n15984 imp/extract/imp_c9.extract\n25929 imp/extract/imp_c10.extract\n25705 imp/extract/imp_c11.extract\n23941 imp/extract/imp_c12.extract\n13520 imp/extract/imp_c13.extract\n16043 imp/extract/imp_c14.extract\n15787 imp/extract/imp_c15.extract\n17556 imp/extract/imp_c16.extract\n19684 imp/extract/imp_c17.extract\n9922 imp/extract/imp_c18.extract\n17248 imp/extract/imp_c19.extract\n13402 imp/extract/imp_c20.extract\n6694 imp/extract/imp_c21.extract\n10350 imp/extract/imp_c22.extract\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "%%bash\nhead -n2 imp/extract/imp_c22.extract # -> for plink \necho \"\"\nhead -n2 imp/extract/imp_c22.bothIDs.extract # -> for matching", "metadata": {"trusted": true, "tags": []}, "execution_count": 29, "outputs": [{"name": "stdout", "text": "22:17060409_TTTTG_T\n22:18223198_CT_C\n\n22:17054103:G:A rs4008588\n22:17054720:T:C rs9605903\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "%%bash\nn=$(cat imp/extract/imp_c*.bothIDs.extract | wc -l)\nnebb=$(cat ebb.variants.1 | wc -l)\n\necho \"$n out of $nebb ($((100*n/nebb)) %) weights' variants in UKB imputed data (INFO > 0.8 & MAF > 0.01)\"", "metadata": {"trusted": true, "tags": []}, "execution_count": 34, "outputs": [{"name": "stdout", "text": "469881 out of 473982 (99 %) weights' variants in UKB imputed data (INFO > 0.8 & MAF > 0.01)\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "!dx upload -r imp/extract/ --dest vasilis/data/ebb/extract/", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "# !dx upload -r imp/extract/*bothIDs.extract --dest vasilis/data/ebb/extract/", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ndx upload mwas_01_extract.sh --dest vasilis/SAK_scripts/", "metadata": {"trusted": true, "tags": []}, "execution_count": 133, "outputs": [{"name": "stdout", "text": "ID                                file-J60P7QjJZB79k2xq9v6gqz7Q\nClass                             file\nProject                           project-GfvP6PQJZB72v2Vk348Bb2yg\nFolder                            /vasilis/SAK_scripts\nName                              mwas_01_extract.sh\nState                             closing\nVisibility                        visible\nTypes                             -\nProperties                        -\nTags                              -\nOutgoing links                    -\nCreated                           Mon Feb  2 18:58:11 2026\nCreated by                        vasilisraptis\n via the job                      job-J607PJ0JZB75g08bgjY771xk\nLast modified                     Mon Feb  2 18:58:12 2026\nMedia type                        \narchivalState                     \"live\"\ncloudAccount                      \"cloudaccount-dnanexus\"\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "%%bash\nbgen_dir=\"/Bulk/Imputation/UKB imputation from genotype\"\ndatadir=\"vasilis/data/ebb\"\ndest=\"vasilis/data/ebb/imp_bed/\"\n\nfor CHR in {1..2}; do\n    dx run swiss-army-knife \\\n        -iin=\"vasilis/SAK_scripts/mwas_01_extract.sh\" \\\n        -iin=\"${bgen_dir}/ukb22828_c${CHR}_b0_v3.bgen\" \\\n        -iin=\"${bgen_dir}/ukb22828_c${CHR}_b0_v3.sample\" \\\n        -iin=\"${datadir}/wb.eids\" \\\n        -iin=\"${datadir}/extract/imp_c${CHR}.extract\" \\\n        -icmd=\"sh mwas_01_extract.sh ${CHR}\" \\\n        --tag=\"ext_${CHR}\" \\\n        --instance-type \"mem1_ssd1_v2_x72\" \\\n        --destination=\"${dest}\" \\\n        --brief --yes --priority high\ndone\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 22, "outputs": [{"name": "stdout", "text": "job-J60xGx0JZB7F6YGYY5FvzK1P\njob-J60xGy0JZB76F8pk230GFZv8\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}]}